<p>Provenance; Common Workflow Language; CWL; Research Object; RO; BagIt; Interoperability; Scientific Workflows; Containers</p>
<h1 id="sec:introduction">Introduction</h1>
<p>Out of the many big data domains, genomics is considered <em>“the most demanding”</em> with respect to all stages of the data lifecycle - from acquisition, storage, distribution and analysis <span class="citation">(Stephens et al. 2015)</span>. As genomic data is growing at an unprecedented rate due to improved sequencing technologies and reduced cost, it is currently challenging to analyse the data at a rate matching its production. With data growing exponentially in size and volume, the practice to perform computational analyses using <em>workflows</em> has overtaken more traditional research methods using ad-hoc scripts which were the typical modus operandi over the last few decades <span class="citation">(Atkinson et al. 2017; Spjuth et al. 2015)</span>. Scientific workflow design and management has become an essential part of many computationally driven data-intensive analyses enabling Automation, Scaling, Adaptation and Provenance support (ASAP)<span class="citation">(Cuevas-Vicenttín et al. 2012)</span>. Increased use of workflows has driven rapid growth in the number of computational data analysis WMSs, with hundreds of heterogeneous approaches now existing for workflow specification and execution <span class="citation">(Common Workflow Language project 2018)</span>. There is an urgent need for a common format and standard to define workflows and enable sharing of analysis results using a given workflow environment.</p>
<p>[contributions] The contribution of this paper is fourfold:</p>
<ul>
<li><p>We have gathered best-practice recommendations from the existing literature, and reflect on the various authors’ experiences with workflow managements systems and especially with regards to factors to consider when a computational analysis is designed, executed and shared.</p></li>
<li><p>Combining the above with our own experiences from empirical studies <span class="citation">(Kanwal et al. 2017; Wolstencroft et al. 2013; Möller et al. 2017; Alterovitz 2019)</span>, we define a set of hierarchical levels of provenance tracking and method sharing where the highest level represent complete understanding of the shared resources supported by reproducibility and re-use of the methods from the lower levels.</p></li>
<li><p>Building on this provenance hierarchy, we define <em>CWLProv</em> for the methodical representation of artefacts associated with a given workflow enactment associated with any study involving computational data-intensive analysis.</p></li>
<li><p>Finally, we demonstrate the utilisation of <em>CWLProv</em> by extending an existing workflow execution engine <em>cwltool</em> <span class="citation">(Amstutz et al. 2018)</span> to produce workflow-centric Research Objects generated as a result of a given workflow enactment. We illustrate this through a case study of using workflows designed by external (independent) developers, and subsequently evaluate the interoperability, reproducibility and completeness of the generated <em>CWLProv</em> outcome.</p></li>
</ul>
<p>Common Workflow Language (CWL) <span class="citation">(Amstutz et al. 2016)</span> has emerged as a workflow definition standard designed to enable portability, interoperability and reproducibility of analyses between workflow platforms. CWL has been widely adopted by more than 20 organisations, providing an interoperable bridge overcoming the heterogeneity of workflow environments. Whilst a common standard for workflow definition is an important step towards interoperable solutions for workflow specifications, sharing and publishing the <em>results</em> of these workflow enactments in a common format is equally important. Transparent and comprehensive sharing of experimental designs is critical to establish trust and ensure authenticity, quality and reproducibility of any workflow-based research result. Currently there is no common format defined and agreed upon for interoperable workflow archiving or sharing <span class="citation">(Ivie and Thain 2018)</span>.</p>
<p>In this paper, we utilize open-source standards such as CWL together with related efforts such as Research Objects (ROs) <span class="citation">(Belhajjame et al. 2015)</span>, BagIt <span class="citation">(Network Working Group 2017)</span> and PROV <span class="citation">(Missier, Belhajjame, and Cheney 2013)</span> to define <strong><em>CWLProv</em></strong>, a format for the interoperable representation of a CWL workflow enactment. We focus on production of a workflow-centric executable RO as the final result of a given CWL workflow enactment. This RO is equipped with the artefacts used in a given execution including the workflow inputs, outputs and, most importantly, the retrospective provenance. This approach enables the complete sharing of a computational analysis such that any future CWL-based workflow can be re-run given the best practices discussed later for software environment provision are followed.</p>
<p>The concept of workflow-centric ROs has been previously considered <span class="citation">(Belhajjame et al. 2015; Hettne et al. 2014; Belhajjame et al. 2012)</span> for structuring the analysis methods and aggregating digital resources utilized in a given analysis. The generated ROs in these studies typically aggregate data objects, example inputs, workflow specifications, attribution details, details about the execution environment amongst various other elements. These previous efforts were largely tied to a single platform or a single Workflow Management System (WMS). <em>CWLProv</em> aims to provide a platform-independent solution for workflow sharing, enactment and publication. All the standards and vocabularies used to design <em>CWLProv</em> have an overarching goal to support a domain-neutral and interoperable solution (detailed in Section <strong></strong>).</p>
<p>The contribution of this work are summarized and listed in the <strong>Key Points</strong> section and the remainder of this paper is structured as follows. In Section <strong></strong> we discuss the key concepts and related work followed by a summary of the published best-practices and recommendations for workflow representation and sharing in Section <strong></strong>. This section also details the hierarchical provenance framework that we define to provide a principled approach for provenance capture and method sharing. Section <strong></strong> introduces <em>CWLProv</em> and outlines its format, structure and the details of the standards and ontologies it utilizes. Section <strong></strong> presents the implementation details of <em>CWLProv</em> using <em>cwltool</em> <span class="citation">(Amstutz et al. 2018)</span> and Section <strong></strong> demonstrates and evaluates the implemented module for three existing workflow case studies. We discuss the challenges of interoperable workflow sharing and the limitations of the proposed solution listing several possible future research directions in Section <strong></strong> before finally drawing conclusions on the work as a whole in Section <strong></strong>.</p>
<h1 id="sec:background">Background and Related Work</h1>
<p>This work draws upon a range of topics as <em>Provenance</em> and <em>Interoperability</em>. We define these here to provide better context for the reader.</p>
<h2 id="sec:provenance">Provenance</h2>
<p>A number of studies have advocated the need for complete provenance tracking of scientific workflows to ensure transparency, reproducibility, analytic validity, quality assurance and attribution of (published) research results <span class="citation">(Herschel, Diestelkämper, and Ben Lahmar 2017)</span>. The term <em>Provenance</em> is defined by World Wide Web Consortium (W3C) <span class="citation">(Moreau, Missier, Belhajjame, et al. 2013)</span> as:</p>
<blockquote>
<p><em>“Provenance is information about entities, activities, and people involved in producing a piece of data or thing, which can be used to form assessments about its quality, reliability or trustworthiness.”</em></p>
</blockquote>
<p>Provenance for workflows is commonly divided into the following three categories: <em>Retrospective Provenance</em>; <em>Prospective Provenance</em> and <em>Workflow Evolution</em>. <em>Retrospective Provenance</em> refers to the detailed record of the implementation of a computational task including the details of every executed process together with comprehensive information about the execution environment used to derive a specific product. <em>Prospective Provenance</em> refers to the ‘recipes’ used to capture a set of computational tasks and their order, e.g. the workflow specification <span class="citation">(Clifford et al. 2008)</span>. This is typically given as an abstract representation of the steps (tools/data analysis steps) that are necessary to create a particular research output, e.g. a data artefact. <em>Workflow Evolution</em> refers to tracking of any alteration in the existing workflow resulting in another version of the workflow that may produce either the same or different resultant data artefacts <span class="citation">(Casati et al. 1998)</span>. In this work, our focus is mainly on improving representation and capture of <em>Retrospective Provenance</em>.</p>
<h2 id="sec:interoperability">Interoperability</h2>
<p>The concept of interoperability varies in different domains. Here we focus on <em>computational interoperability</em> defined as:</p>
<blockquote>
<p><em>The ability of two or more components or systems to exchange information and to use the information that has been exchanged</em> <span class="citation">(“Interoperability | Definition of Interoperability in English by Oxford Dictionaries”)</span>.</p>
</blockquote>
<p>The focus of this study is to propose and devise methods to achieve <em>syntactic</em>, <em>semantic</em> and <em>pragmatic</em> interoperability as defined in Levels of Conceptual Interoperability Model (LCIM)<span class="citation">(Tolk, n.d.)</span>. <em>Syntactic</em> interoperability is achieved when a common data format for information exchange is unambiguously defined. The next level of interoperability, referred to as <em>semantic</em> interoperability, is reached when the content of the actual information exchanged is unambiguously defined. Once there is an agreement about the format and content of the information, <em>pragmatic</em> interoperability is achieved when the context, application and use of the shared information and data exchanged is also unambiguously defined. In the section <strong></strong>, we relate these general definitions to specific workflow applications with respect to workflow-centric ROs and describe to what extent these interoperability requirements are addressed.</p>
<h2 id="sec:relwork">Related Work</h2>
<p>We focus on relevant studies and efforts trying to resolve the issue of availability of required resources used in a given computational analysis. In addition, we cover efforts directed towards provenance capture of workflow enactments. As these concepts have been around for a considerable time, we restrict our attention to scientific workflows and studies related to the bioinformatics domain.</p>
<h3 id="workflow-software-environment-capture"> Workflow Software Environment Capture</h3>
<p><em>Freezing</em> and packaging the run-time environment to encompass all the software components and their dependencies used in an analysis is a recommended and widely adopted practice <span class="citation">(Cohen-Boulakia et al. 2017)</span> especially after use of cloud computing resources where images and snapshots of the cloud instances are created and shared with fellow researchers <span class="citation">(Howe 2012)</span>. Nowadays, preservation and sharing of the software environment e.g. in open access repositories, is becoming a regular practice in the workflow domain as well. Leading platforms managing infrastructure and providing cloud computing services and configuration on demand include DigitalOcean <span class="citation">(“DigitalOcean - Cloud Computing, Simplicity at Scale” 1AD–12AD)</span>, Amazon Elastic Compute Cloud <span class="citation">(“Amazon Ec2” 1AD–12AD)</span>, Google Cloud Platform <span class="citation">(“Google Cloud Including Gcp &amp; G Suite Url = https://cloud.google.com/, Note = Accessed 01 Dec 2018, Urldate = 01/12/2018,” n.d.)</span> and Microsoft Azure <span class="citation">(“Microsoft Azure Cloud Computing Platform &amp; Services” 1AD–12AD)</span>. The instances launched on these platforms can be saved as snapshots and published with an analysis study to later re-create an instance representing the computing state at analysis time.</p>
<p>Using <em>“System-wide packaging”</em> for data-driven analyses, although simplest on part of the workflow developers and researchers, has its own caveats. One of the notable issue is the size of the snapshot as it captures everything in an instance at a given time, hence the size can range from few gigabytes to many terabytes. To distribute research software and share execution environments, various light-weight and container-based virtualisation and package managers are emerging, including: Docker, Singularity, Debian Med and Bioconda.</p>
<p><em>Docker</em><span class="citation">(Merkel 2014)</span> is a lightweight container-based virtualisation technology that facilitates the automation of application development by archiving software systems and environment to improve portability of the applications on many common platforms including Linux, Microsoft windows, Mac OS X and cloud instances. <em>Singularity</em><span class="citation">(Kurtzer, Sochat, and Bauer 2017)</span> is also a cross-platform open source container engine specifically supporting High Performance Computing (HPC) resources. An existing Docker format software image can be imported and used by the Singularity container engine. <em>Debian Med</em> <span class="citation">(Möller et al. 2017)</span> contribute packages of medical practice and biomedical research to the Debian Linux distribution, lately also including workflows <span class="citation">(Möller et al. 2017)</span>. <em>Bioconda</em><span class="citation">(Grüning et al. 2018)</span> packages, based on the an open source package manager Conda <span class="citation">(“Conda Conda Documentation”)</span>, are available for Mac OS X and Linux environments, directing towards availability and portability of software used in the life science domain.</p>
<h3 id="datamethod-preservation-aggregation-sharing"> Data/Method Preservation, Aggregation &amp; Sharing</h3>
<p>Preserving and sharing only the software environment is not enough to verify results of any computational analysis or re-use the methods used (e.g. workflows) with a different dataset. It is also necessary to share other details including data (example or the original), scripts, workflow files, input configuration settings, the hypothesis of the experiment and any/all trace/logging information related to “what happened”, i.e. the retrospective provenance of the actual workflow enactment. The publishing of resources to improve state of scholarly publications is now supported by various online repositories, including Zenodo <span class="citation">(“Zenodo - Research. Shared.”)</span>, GitHub <span class="citation">(“GitHub”)</span>, myExperiment <span class="citation">(Goble et al. 2010)</span> and Figshare <span class="citation">(“Figshare - Credit for All Your Research”)</span>. These repositories facilitate collaborative research, in addition to public sharing of source code and the results of a given analysis. There is however no agreed format that must be followed when someone shares artefacts associated with an analysis. As a result, the quality of the shared resources can range from a highly annotated, properly documented and complete set of artefacts, to raw data with undocumented code and incomplete information about the analysis as a whole. Individual organisations or groups might provide a set of “recommended practices”, e.g. in readme files, to attempt to maintain the quality of shared resources. The initiative <em>Code as a Research Object</em> <span class="citation">(“Code as a Research Object”)</span> is a joint project between Figshare, GitHub and Mozilla Science Lab <span class="citation">(“Mozilla Science” 1AD–12AD)</span> and aims to archive any GitHub code repository to Figshare and produce a Digital Object Identifier (DOI) to improve the discovery of resources<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a>.</p>
<p>Reprozip<span class="citation">(Chirigati et al. 2016)</span> aims to resolve portability issues by identifying and packaging all dependencies in a self-contained package which when unpacked and executed on another system (with Reprozip installed) should reproduce the methods and results of the analysis. Each package also contains a human readable configuration file containing provenance information obtained by tracing system calls during system execution. The corresponding provenance trace is however not formatted using existing open standards established by the community. Several platform-dependent studies have been targeted towards extensions to existing standards by implementing the Research Object model and improving aggregation of resources. <span class="citation">Belhajjame et al. (2015)</span> proposed the application of ROs to develop workflow-centric ROs containing data and metadata to support the understandability of the utilized methods (in this case workflow specifications). They explored five essential requirements to workflow preservation and identified data and metadata that could be stored to satisfy the said requirements. These requirements include providing example data, preserving workflows with provenance traces, annotating workflows, tracking the evolution in workflows and packaging the auxiliary data and information with workflows. They proposed extensions to existing ontologies such as Object Reuse and Exchange (ORE), the Annotation Ontology (AO) and PROV-O, with four additional ontologies to represent workflow specific information. However, as stated in the paper, the scope of the proposed model at that time was not focused on interoperability of heterogeneous workflows as it was demonstrated for a workflow specific to Taverna WMS using myExperiment, which makes it quite platform-dependent.</p>
<p>A domain-specific solution is proposed by <span class="citation">Gomez-Perez, Palma, and Garcia-Silva (2017)</span> by extending the RO model to equip workflow-centric ROs with information catering for the specific needs of the Earth Science community, resulting in enhanced discovery and reusability by experts. They demonstrated that the principles of ROs can support extensions to generate aggregated resources leveraging domain specific knowledge. <span class="citation">Hettne et al. (2014)</span> used three genomic workflow case studies to demonstrate the utilisation of ROs to capture methods and data supporting querying and useful extraction of information about the scientific investigation under observation. The solution was tightly coupled with the Taverna WMS and hence if shared, would not be reproducible outside of the Taverna environment. Other notable efforts to use ROs for workflow preservation and method aggregation include <span class="citation">(Wolstencroft et al. 2013)</span> in systems biology, <span class="citation">(Custovic et al. 2015)</span> in clinical settings and <span class="citation">(Alterovitz 2019)</span> in precision medicine.</p>
<h3 id="provenance-capture-standardization"> Provenance Capture &amp; Standardization</h3>
<p>A range of standards for provenance representation have been proposed. Many studies have emphasized the need for provenance focusing on aspects such as scalability, granularity, security, authenticity, modelling and annotation <span class="citation">(Herschel, Diestelkämper, and Ben Lahmar 2017)</span>. They identify the need to support standardized dialogues to make provenance interoperable. Many of these were used as inputs to initial attempts at creating a standard Provenance Model to tackle the often inconsistent and disjointed terminology related to provenance concepts. This ultimately resulted in the specification of the <em>Open Provenance Model</em> (OPM)<span class="citation">(Moreau et al. 2008)</span> together with an open-source model for the governance of OPM <span class="citation">(Moreau et al. 2009)</span>. Working towards similar goals of interoperability and standardization of provenance for web technologies, the World Wide Web Consortium (W3C) Provenance Incubator Group <span class="citation">(“W3C Provenance Incubator Group Wiki - Xg Provenance Wiki”)</span> and the authors of OPM together set the fourth provenance challenge at the International Provenance and Annotation Workshop, 2010 (IPAW’10) that later resulted in <em>PROV</em>, a family of documents serving as the conceptual model for provenance capture, its representation, sharing and exchange over the Web <span class="citation">(Moreau et al. 2015)</span> regardless of the domain or platform. Since then, a number of studies have proposed extensions to this domain-neutral standard. The model is general enough to be adapted to any field and flexible enough to allow extensions for specialized cases.</p>
<p><span class="citation">Michaelides et al. (2016)</span> presented a domain-specific PROV-based solution for retrospective provenance to support portability and reproducibility of a statistical software suite. They captured the essential elements from the log of a workflow enactment and represented them using an intermediate notation. This representation was later translated to PROV-N and used as the basis for the PROV Template System. A Linux specific system provenance approach was proposed in <span class="citation">(Pasquier et al. 2017)</span> where they demonstrated retrospective provenance capture at the system level. Another project <em>UniProv</em> is working to extract information from Unicore middleware and transform it into a PROV-O representation to facilitate the back-tracking of workflow enactments <span class="citation">(Giesler et al. 2017)</span>. Other notable domain-specific efforts leveraging the established standards to record provenance and context information are <em>PROV-man</em> <span class="citation">(Benabdelkader, VanKampen, and Olabarriaga 2015)</span>, PoeM <span class="citation">(Gaignard, Skaf-Molli, and Bihouée 2016)</span> and micropublications <span class="citation">(Clark, Ciccarese, and Goble 2014)</span>. Platforms such as VisTrails and Taverna have built in retrospective provenance support. <em>Taverna</em> <span class="citation">(Wolstencroft et al. 2013)</span> implements an extensive provenance capture system <em>TavernaProv</em><span class="citation">(Soiland-Reyes, Alper, and Goble 2016)</span>, utilising both PROV ontologies as well as ROs aggregating the resources used in an analysis. <em>VisTrails</em><span class="citation">(Freire and Silva 2012)</span> is an open source project supporting platform-dependent provenance capture, visualisation and querying for extraction of required information about a workflow enactment. <span class="citation">(Chirigati et al. 2016)</span> provide an overview of PROV terms and how they can be translated from the VisTrails schema and serialized to PROV-XML. <em>WINGS</em><span class="citation">(Gil et al. 2011)</span> can report fine-grained workflow execution provenance as Linked Data using the OPMW ontology <span class="citation">(Garijo, Gil, and Corcho 2017)</span>, which builds on both PROV-O and OPM.</p>
<p>All these efforts are fairly recent and use a standardized approach to provenance capture and hence are relevant to our work on the capture of retrospective provenance. However, our aim is a domain-neutral and platform-independent solution that can be easily adapted for any domain and shared across different platforms and operating systems.</p>
<p>As evident from the literature, there are efforts in progress to resolve the issues associated with effective and complete sharing of computational analysis including both the results and provenance information. These studies range from highly domain-specific solutions and platform-dependent objects to open source flexible interoperable standards. CWL has widespread adoption as a workflow definition standard, hence is an ideal candidate for portable workflow definitions. The next section investigates existing studies focused on workflow-centric science, and summarises best practice recommendations put forward in these studies. From this we define a hierarchical provenance and resource sharing framework.</p>
<p>[t!] <img src="images/recommendations2.png" alt="image" /></p>
<p><span>p<span>1.4cm</span> L L</span> &amp; <span><strong>Recommendations</strong></span> &amp; <span><strong>Justifications</strong></span><br />
R1 &amp; Save and share all parameters used for each software executed in a given workflow (including default values of parameters used) <span class="citation">(Nekrutenko and Taylor 2012; Garijo et al. 2013; Garijo, Gil, and Corcho 2017; Sandve et al. 2013)</span>. &amp; Impacts on reproducibility of results since different inputs and configurations of the software can produce different results. Different versions of a tool might upgrade the default values of the parameters.<br />
R2 &amp; Avoid manual processing of data and if using <em>shims</em> <span class="citation">(Mohan, Lu, and Kotov 2014)</span> then make these part of the workflow to fully automate the computational process <span class="citation">(Nekrutenko and Taylor 2012; Sandve et al. 2013)</span>. &amp; This ensures the complete capture of the computational process without broken links so that the analysis can be executed without need for performing manual steps.<br />
R3 &amp; Include intermediate results where possible when publishing an analysis <span class="citation">(Garijo et al. 2013; Garijo, Gil, and Corcho 2017; Sandve et al. 2013)</span>. &amp; Intermediate data products can be used to inspect and understand shared analysis when re-enactment is not possible.<br />
R4 &amp; Record the exact software versions used <span class="citation">(Nekrutenko and Taylor 2012; Sandve et al. 2013)</span>. &amp; This is necessary for reproducibility of results as different software versions can produce different results.<br />
R5 &amp; If using public data (reference data, variant databases), then it is necessary to store and share the actual data versions used <span class="citation">(Spjuth et al. 2015; Kanwal et al. 2017; Nekrutenko and Taylor 2012; Sandve et al. 2013)</span> . &amp; This is needed as different versions of data, e.g. human reference genome or variant databases, can result in slightly different results for the same workflow.<br />
R6 &amp; Workflows should be well-described, annotated and offer associated metadata. Annotations such as user contributed tags and versions should be assigned to workflows and shared when publishing the workflows and associated results <span class="citation">(Belhajjame et al. 2015; Belhajjame et al. 2012; Garijo, Gil, and Corcho 2017; Littauer et al. 2012; Stodden et al. 2016)</span> . &amp; Metadata and annotations improve the understandability of the workflow, facilitate independent re-use by someone skilled in the field, make workflows more accessible and hence promote the longevity of the workflows.<br />
R7 &amp; Use and store stable identifiers for all artefacts including the workflow, the datasets and the software components <span class="citation">(Littauer et al. 2012; Stodden et al. 2016)</span>. &amp; Identifiers play an important role in the discovery, citation and accessibility of resources made available in open-access repositories.<br />
R8 &amp; Share the details of the computational environment <span class="citation">(Belhajjame et al. 2015; Kanwal et al. 2017; Stodden et al. 2016)</span> . &amp; Such details support requirements analysis before any re-enactment or reproducibility is attempted.<br />
R9 &amp; Share workflow specifications/descriptions used in the analysis <span class="citation">(Belhajjame et al. 2015; Garijo et al. 2013; Garijo, Gil, and Corcho 2017; Stodden et al. 2016; Stodden and Miguez 2014)</span>. &amp; The same workflow specifications can be used with different datasets thereby supporting re-usability.<br />
R10 &amp; Aggregate the software with the analysis and share this when publishing a given analysis <span class="citation">(Belhajjame et al. 2015; Kanwal et al. 2017; Stodden et al. 2016; Stodden and Miguez 2014; Garijo, Gil, and Corcho 2017)</span>. &amp; Making software available reduces dependence on third party resources and as a result minimizes <em>workflow decay</em> <span class="citation">(Zhao et al. 2012)</span>.<br />
R11 &amp; Share raw data used in the analysis <span class="citation">(Belhajjame et al. 2015; Garijo et al. 2013; Garijo, Gil, and Corcho 2017; Stodden et al. 2016; Stodden and Miguez 2014)</span>. &amp; When someone wants to validate published results, availability of data supports verification of claims and hence establishes trust in the published analysis<br />
R12 &amp; Store all attributions related to data resources and software systems used <span class="citation">(Garijo, Gil, and Corcho 2017; Stodden and Miguez 2014)</span>. &amp; Accreditation supports proper citation of resources used.<br />
R13 &amp; Workflows should be preserved along with the provenance trace of the data and results <span class="citation">(Belhajjame et al. 2015; Belhajjame et al. 2012; Garijo, Gil, and Corcho 2017; Sandve et al. 2013; Stodden and Miguez 2014)</span>. &amp; A provenance trace provides a historical view of the workflow enactment, enabling end users to better understand the analysis retrospectively<br />
R14 &amp; Data flow diagrams of the computational analysis using workflows should be provided <span class="citation">(Kanwal et al. 2017; Garijo et al. 2013)</span>. &amp; These diagrams are easy to understand and provide a human readable view of the workflow.<br />
R15 &amp; Open source licensing for methods, software, code, workflows and data should be adopted instead of proprietary resources <span class="citation">(Kanwal et al. 2017; Garijo et al. 2013; Sandve et al. 2013; Stodden et al. 2016; Stodden and Miguez 2014; Gymrek and Farjoun 2016)</span>. &amp; This improve availability and legal re-use of the resources used in the original analysis, while restricted licenses would hinder reproducibility.<br />
R16 &amp; Data, code and all workflow steps should be shared in a format that others can easily understand preferably in a system neutral language <span class="citation">(Belhajjame et al. 2015; Garijo et al. 2013; Gymrek and Farjoun 2016)</span>. &amp; System neutral languages help achieve interoperability and make an analysis understandable.<br />
R17 &amp; Promote easy execution of workflows without making significant changes to the underlying environment <span class="citation">(Spjuth et al. 2015)</span>. &amp; In addition to helping reproducibility, this enables adapting the analysis methods to other infrastructures and improves workflow portability.<br />
R18 &amp; Information about compute and storage resources should be stored and shared as part of the workflow <span class="citation">(Kanwal et al. 2017)</span>. &amp; Such information can assist users in estimating the required resources needed for an analysis and thereby reduce the amount of failed executions.<br />
R19 &amp; Example input and sample output data should be preserved and published along with the workflow-based analysis <span class="citation">(Belhajjame et al. 2015; Zhao et al. 2012)</span>. &amp; This information enables more efficient test runs of an analysis to verify and understand the methods used.<br />
</p>
<p>This list is not exhaustive, other studies have identified separate issues (e.g. lab work provenance and data security) that are beyond the scope of this work.</p>
<h1 id="sec:levels">Levels of Provenance and Resource Sharing</h1>
<p>Various studies have empirically investigated the role of automated computational methods in the form of workflows and published best practice recommendations to support workflow design, preservation, understandability and re-use. We summarise a number of these recommendations and the their justifications in Table [tab:recommendation:wide], where each recommendation addresses specific requirement of workflow design and sharing. These recommendations can be clustered into broad themes as shown in Figure [fig:recommendationclasses]. This classification can be in more than one way e.g. according to how these recommendations are supporting each FAIR dimension <span class="citation">(Wilkinson et al. 2016)</span>. In this study, we have focused on categories with respect to workflow design, prospective provenance, data sharing, retrospective provenance, the computational environment required/used for an analysis and lastly better findability and understandability of all shared resources.</p>
<p>Sharing <em>“all artefacts”</em> from a computational experiment (following all recommendations and best practices) is a demanding task without any informed guidance. It requires consolidated understanding of the impact of the many different artefacts involved in that analysis. This places extra efforts on workflow designers, (re)-users, authors, reviewers and expectations on the community as a whole. Given the numerous WMS and differences in how each system deals with provenance documentation, representation and sharing of these artefacts, the granularity of provenance information preserved will vary for each workflow definition approach. Hence, devising one universal but technology-specific solution for provenance capture and the related resource sharing is impossible. Instead we propose a generic framework of provenance in Figure [fig:levels] that all WMSs can benefit from and conform to with minimum technical overheads.</p>
<p>The recommendations in Table [tab:recommendation:wide] aid in our understanding to define this framework by classifying the granularity of the provenance and related artefacts where the uppermost level exhibits comprehensive, reproducible, understandable and provenance-rich computational experiment sharing. The purpose of this framework is threefold. First, because of its generic nature it brings the uniformity in the provenance granularity across various WMS belonging to different workflow definition approaches. Second, it provides comprehensive and well-defined guidelines that can be used by the researchers to conduct principled analysis of the provenance of any published study. Third, due to its hierarchical nature, the framework can be leveraged by the workflow authors to progress incrementally towards the most transparent workflow-centric analysis. Overall, this framework will help achieve a uniform level of provenance and resource sharing with a given workflow-centric analysis guaranteed to fulfill the respective provenance applications.</p>
<p>Our proposed provenance levels are ordered from low granularity to higher degrees of specificity. In brief, <strong>Level 0</strong> is unstructured information about the overall workflow enactment, <strong>Level 1</strong> adds structured retrospective provenance, access to primary data and executable workflows, <strong>Level 2</strong> enhances the white-box provenance for individual steps, and <strong><em>Level 3</em></strong> adds domain-specific annotations for improved understanding. These levels are described in the following sub-sections and mapped to the requirements in Table [tab:recommendation:wide] that these levels aim to satisfy.</p>
<p><img src="images/ProvenanceLevels" alt="image" /></p>
<h2 id="sec:level0">Level 0</h2>
<p>To achieve this level, researchers should share the workflow specifications, input parameters used for a given workflow enactment, raw logs and output data preferably through an open-access repository. This is the least information that could be shared without putting any extra efforts to support seamless reuse or understandability of a given analysis. The artefacts shared at this level would only require uploading of the associated resources to a repository without necessarily providing any supporting metadata or provenance information. Information captured at <em>Level 0</em> is the bare minimum that can be used for result interpretation.</p>
<p>Workflow definitions based on <em>Level 0</em> can also potentially be re-purposed for other analyses. As argued by Ludäscher, a well-written scientific workflow and its graphical representation is itself a source of prospective provenance giving user an idea of the steps taken and data produced <span class="citation">(Ludäscher 2016)</span>. Therefore a well-described workflow specification indirectly provides prospective provenance without aiming for it. In addition to the textual workflow specification, its graphical representation should also be shared if available for better understandability fulfilling <em>R14-diagram</em>. At this level, reproducing the workflow would only be possible if the end-user devotes extra efforts to understand the shared artefacts and carefully recreate the execution environment. As open access journals frequently require availability of methods and data, many published studies now share workflow specifications and optionally the outputs thereby achieving <em>Level 0</em> and specifically satisfying <em>R1-parameters</em> and <em>R9-workflow</em> (Table [tab:recommendation:wide]). In addition, the resources shared should have open licence starting from <em>Level 0</em> and this practice proposed by <em>R15-open-source</em> should be adopted at each higher level.</p>
<h2 id="sec:level1">Level 1</h2>
<p>At <em>Level 1</em>, <em>R4-sw-version</em>, <em>R5-data-version</em>, <em>R12-attribution</em> and <em>R13-provenance</em> should be satisfied by providing retrospective provenance of the workflow enactment - i.e. a structured representation of machine readable provenance which can answer questions such as “what happened”, “when happened”, “what was executed”, “what was used”, “who did this” and “what was produced”. Seamless re-enactment of the workflow should be supported at this level. This is only possible when along with provenance information, <em>R8-environment</em> and <em>R10-software</em> is satisfied by potentially packaging the software environment for analysis sharing or there is enough information about the software environment that guide the user to reliably re-enact the workflow. Hence <em>R17-executable</em> should be satisfied making it possible for the end users to re-enact the shared analyses without making major changes to the underlying software environment.</p>
<p>In addition to the software availability and retrospective provenance, access to input data should also be provided fulfilling <em>R11-raw-data</em>. This data can be used to re-enact the published methods or utilized in a different analysis, e.g. for performance comparison of methods. At <em>Level 1</em>, it is preferable to provide content-addressable data artefacts such as input, output and intermediate files, avoiding local paths and file names to make a given workflow executable outside its local environment. The intermediate data artefacts should also be provided to facilitate inspection of all step results, hence satisfying <em>R3-intermediate</em>. All resources, including workflow specifications and provenance, should be shared in a format that is understandable across platforms, preferably in a technology-neutral language as proposed by <em>R16-format</em>.</p>
<p>While software and data can be digitally captured, the hardware and infrastructure requirements also need to be captured to fulfill <em>R18-resource-use</em>. This kind of information can naturally vary widely with runtime environments, architectures and data sizes <span class="citation">(Bubak et al. 2013)</span>, as well as rapidly becoming outdated as hardware and cloud offerings evolve. Nevertheless a snapshot of the workflow’s overall execution resource usage for an actual run can be beneficial to give a broad overview of the requirements, and can facilitate cost-efficient re-computation by taking advantage of spot-pricing for cloud resources <span class="citation">(Angiuoli et al. 2011)</span>.</p>
<h2 id="sec:level2">Level 2</h2>
<p>It is a common practice in scientific workflows to modularize the workflow specifications by separating the related tasks into “sub-workflows” or “nested workflows” <span class="citation">(Cohen-Boulakia et al. 2017)</span> to be incorporated and used in other workflows or be assigned to compute and storage resources in case of distributed computing <span class="citation">(Chen and Deelman 2011)</span>. These modular solutions promote understanding and re-usability of the workflows as researchers are inclined to use these modules instead of workflow as whole for their own computational experiments. An example of a sub-workflow is the mandatory “pre-processing” <span class="citation">(“GATK | Bp Doc #11165 | Data Pre-Processing for Variant Discovery” 2018)</span> needed for the Genome Analysis ToolKit (GATK) best practice pipelines used for genomic variant calling. These steps can be separated into a sub-workflow to be used before any variant calling pipeline, be it somatic or germline.</p>
<p>At <em>Level 1</em>, retrospective provenance is coarse grained and as such, there is no distinction between workflows and their sub-workflows. Ludäscher <span class="citation">(Ludäscher 2016)</span> distinguishes workflow provenance between <em>black-box</em> and database provenance as <em>white-box</em>. The reasoning behind this distinction is that often the steps in a workflow, especially those based on graphical user interface-based platforms, provide levels of abstraction/obscurity to the actual tasks being implemented. In our previous work we used an empirical case study to demonstrate that declarative approaches to workflow definition resulted in transparent workflows with the least number of assumptions <span class="citation">(Kanwal et al. 2017)</span>. This resolves the black box/white box issue to some extent, but to further support research transparency, we propose to share retrospective provenance logs for each nested/sub-workflow making the details of a workflow enactment as explicit as possible and moving a step closer to <em>white-box</em> provenance. These provenance logs will support the inspection and automatic re-enactment of targeted components of a workflow such as a single step or a sub-workflow individually without necessarily having to re-enact the full analysis. Some existing make-like systems such as Snakemake support partial re-enactments but typically rely on fixed file paths for input data and require manual intervention to provide the specific directory structure. With detailed provenance logs and the corresponding content-addressable data artefacts, the partial re-runs can be achieved with automatic generation of input configuration setting.</p>
<p>In addition, we propose to include <em>permalinks</em> at <em>Level 2</em> to identify the workflows and their individual steps which facilitates the inspection of each step and aim to improve the longevity of the shared resources, hence supporting <em>R7-identifier</em>. Improving <em>R18-resource-use</em> for <em>Level 2</em> would include resource usage per task execution. Along with execution times this can be useful information to identify bottlenecks in a workflow and for more complex calculations in cost optimization models <span class="citation">(Malawski, Figiela, and Nabrzyski 2013)</span>. At this provenance level resource usage data will however also become more noisy and highly variant on scheduling decisions by the workflow engine, e.g. sensitivity to cloud instance reuse or co-use for multiple tasks, or variation in data transfers between tasks on different instances. Thus <em>Level 2</em> resource usage information should be further processed with statistical models for it to be meaningful for a user keen to estimate the resource requirement for re-enactment of a given analysis.</p>
<h2 id="sec:level3">Level 3</h2>
<p>Levels 0-2 are generic and domain-neutral, and can apply to any scientific workflow. However, domain-specific information/metadata about data and processes plays an important role in better understanding of the analysis and exploitation of provenance information, e.g. for meaningful queries to extract information to the domain under consideration <span class="citation">(Alper et al. 2018; Gaignard et al. 2014)</span>. Addition of domain specific metadata e.g. file formats, user-defined tags and other annotations to generic retrospective provenance can improve the <em>white-boxness</em> by providing domain context to the analysis as described in <em>R6-annotations</em>. Annotations can range from adding textual description and tags to marking data with more systematic and well-defined domain-specific ontologies such as EDAM <span class="citation">(Ison et al. 2013)</span> and BioSchemas <span class="citation">(Michel and Community 2018)</span> in the case of bioinformatic workflows. Some studies also propose to provide example or test data sets which eventually helps in analyzing the methods shared and verifying their results (as described in <em>R19-example</em>).</p>
<p>At <em>Level 3</em>, the information from previous levels combined with specific metadata about data artefacts facilitates higher level classification of workflow steps into <em>motifs</em> <span class="citation">(Garijo, Gil, and Corcho 2014)</span> such as data retrieval, pre-processing, analysis and visualisation. This level of provenance, resource aggregation and sharing can provide a researcher-centric view of data and enable users to re-enact a set of steps or full workflow by providing filtered and annotated view of the execution. This can be non-trivial to achieve with mainstream methods of workflow definition and sharing, as it requires guided user annotations with controlled vocabularies, but this can be simplified by reusing related tooling from existing efforts like BioCompute Objects <span class="citation">(Alterovitz 2019)</span> and DataCrate <span class="citation">(Sefton et al. 2018)</span>.</p>
<p>Communicating resource requirements (<em>R18-resource-use</em>) at <em>Level 3</em> would involve domain-specific models for hardware use and cost prediction, as suggested for dynamic cloud costing <span class="citation">(Woods 2018)</span> in <em>BioSimSpace</em> <span class="citation">(“BioSimSpace” 2018)</span>, or predicting assembler and memory settings through machine learning of variables like source biome, sequencing platform, file size, read count and base count in the <em>European Bionformatics Institute (EBI) Metagenomics</em> pipeline <span class="citation">(Mitchell et al. 2018)</span>. For robustness such models typically need to be derived from resource usage across multiple workflow runs with varied inputs, e.g. by a multi-user workflow platform. Taking advantage of <em>Level 3</em> resource usage models might require pre-processing workflow inputs and calculations in an environment like R or Python, and so we recommend that models are provided with separate sidecar workflows for interoperable execution before the main workflow.</p>
<p>By explicit enumeration of the levels of provenance, it should be possible to quantify and directly assess the effort required to re-use a workflow and reproduce experiments directly. Similar effort like <em>5-star Open Data</em> <span class="citation">(Kim and al. 21AD–5AD)</span> strongly advocates open-licensed structured representation, use of stable identifiers for data sharing and following Linked Data principles to cross-relate data. One challenge on achieving the Open Data stars is that it needs tool support during data processing. In our framework we proposed systematic workflow-centric resource sharing using structured Linked Data representation, including recording of the executed data operations. Hence, our effort compliments the already proposed 5-star Open Data principles and contributes to further understanding by sharing the computational method following the same principles.</p>
<p>Requiring researchers to achieve the above defined levels individually is unrealistic without guidance and direct technical support. Ideally, the conceptual meaning of these levels would be translated into a practical solution utilising the available resources. However, given the heterogeneity of workflow definition approaches, it is expected that the proposed framework, when translated into practical solutions, will also naturally result in varying workflow-centric solutions tied to specific WMSs. To support interoperability of the workflow-centric analysis achieving the provenance levels, we propose <strong><em>CWLProv</em></strong>, a format for annotating resource aggregations equipped with retrospective provenance. The next section describes <em>CWLProv</em> and the associated standards that are applied in this process.</p>
<h1 id="sec:CWLProv">CWLProv 0.6.0 and utilized standards</h1>
<p>Here we present <em>CWLProv</em>, a format for the methodical representation of workflow enactment, associated artefacts and capturing and using retrospective provenance information. Keeping in view the recommendations from Table [tab:recommendation:wide] for example <em>R15-open-source</em> and <em>R16-format</em>, we leverage <strong>open-source</strong>, <strong>domain-independent</strong>, <strong>system-neutral</strong>, <strong>interoperable</strong> and most importantly <strong>community-driven</strong> standards as the basis for the design and formatting of reproducible and interoperable workflow-based ROs. The profile description in this section correspond to <em>CWLProv</em> 0.6.0 <span class="citation">(Soiland-Reyes, Khan, and Crusoe 2018)</span>. (see <a href="https://w3id.org/cwl/prov" class="uri">https://w3id.org/cwl/prov</a> for the latest profile).</p>
<h2 id="sec:standards">Applied Standards and Vocabularies</h2>
<p>We follow the recommendation <em>“Reuse vocabularies, preferably standardized ones”</em> <span class="citation">(Lóscio et al. 2017)</span> from best practices associated with data sharing, representation and publication on the web to achieve consensus and interoperability of workflow-based analyses. Specifically we integrate the <em>Common Workflow Language</em> (CWL) for workflow definition, <em>Research Objects</em> (ROs) for resource aggregation and the <em>PROV-Data Model</em> (PROV-DM) to support the retrospective provenance associated with workflow enactment. The key properties and principles of these standards are described below.</p>
<h3 id="common-workflow-language-cwl">Common Workflow Language (CWL)</h3>
<p>Common Workflow Language <span class="citation">(Amstutz et al. 2016)</span> provides declarative constructs for workflow structure and command line tool interface definition. It makes minimal assumptions about base software dependencies, configuration settings, software versions, parameter settings or indeed the execution environment more generally <span class="citation">(Kanwal et al. 2017)</span>. The CWL object model supports comprehensive recording and capture of information for workflow design and execution. This can subsequently be published as structured information alongside any resultant analysis using that workflow.</p>
<p>CWL is a community-driven standard effort that has been widely adopted by many workflow design and execution platforms, supporting interoperability across a set of diverse platforms. Current adopters include Toil, Arvados, Rabix <span class="citation">(Kaushik et al. 2017)</span>, Cromwell <span class="citation">(Voss, Auwera, and Gentry 2017)</span>, REANA, and Bcbio <span class="citation">(Guimera 2012)</span> with implementations for Galaxy, Apache Taverna, and AWE currently in progress.</p>
<p>[t!] <img src="images/twostep" alt="image" /></p>
<p>A workflow in CWL is composed of “steps” where each step refers either to a command line tool (also specified using CWL) or another workflow specification incorporating the concept of “sub-workflows”. Each “step” is associated with “inputs” that are comprised of any data artefact required for the execution of that step (Figure [fig:bwa-mem]). As a result of the execution of each step, “outputs” are produced which can become (part of) “inputs” for the next steps making the execution data-flow oriented. CWL is not tied to a specific operating system or platform which makes it an ideal approach for interoperable workflow definitions.</p>
<h3 id="research-object-ro"> Research Object (RO)</h3>
<p>A Research Object encapsulates all of the digital artefacts associated with a given computational analysis contributing towards preservation of the analysis <span class="citation">(Bechhofer et al. 2013)</span>, together with their metadata, provenance and identifiers.</p>
<p>The aggregated resources can include but are not limited to: input and output data for analysis results validation; computational methods such as command line tools and workflow specifications to facilitate workflow re-enactment; attribution details regarding users; retrospective as well as prospective provenance for better understanding of workflow requirements, and machine-readable annotations related to the artefacts and the relationships between them. The goal of ROs is to make any published scientific investigation and the produced artefacts <em>“interoperable, reusable, citable, shareable and portable”</em>.</p>
<p>The three core principles <span class="citation">(“Researchobject.org”)</span> of the RO approach are to support “Identity”, “Aggregation”, and “Annotation” of research artefacts. They look to enable accessibility of tightly-coupled, interrelated and well-understood aggregated resources involved in a computational analysis as identifiable objects, e.g. using unique (persistent) identifiers such as DOIs and/or ORCIDs. The RO approach is well aligned with the idea of interoperable and platform-independent solutions for provenance capture of workflows because of its domain-neutral and platform-independent nature.</p>
<p>While ROs can be serialized in several different ways, in this work we have reused the BDBag approach based on <em>BagIt</em> (see box), which has been shown to support large-scale workflow data <span class="citation">(Chard et al. 2016)</span>. This approach is also compatible with data archiving efforts from the NIH Data Commons, Library of Congress and the Research Data Alliance. The specialized workflow-centric RO in this study encompasses the components mentioned in the previous paragraph annotated with various targeted tools and a PROV-based <em>Workflow provenance profile</em> to capture the detailed retrospective provenance of the CWL workflow enactment.</p>
<h3 id="prov-data-model-prov-dm">PROV Data Model (PROV-DM)</h3>
<p>The World Wide Web Consortium (W3C) developed <em>PROV</em>, a suite of specifications for unified/interoperable representation and publication of provenance information on the Web. The underlying conceptual PROV Data Model (PROV-DM) <span class="citation">(Moreau, Missier, Belhajjame, et al. 2013)</span> provides a domain-agnostic model designed to capture fundamental features of provenance with support for extensions to integrate domain-specific information (Figure [fig:prov-dm]).</p>
<p>[!b] <embed src="images/key-concepts.pdf" title="fig:" /></p>
<p>We utilize mainly two serialisations of PROV for this study, PROV-Notation (PROV-N) <span class="citation">(Moreau, Missier, Cheney, et al. 2013)</span> and PROV-JSON <span class="citation">(Huynh et al. 2013)</span>. PROV-N is designed to achieve serialisation of PROV-DM instances by formally representing the information using a simplified textual syntax to improve human readability. PROV-JSON is a lightweight interoperable representation of PROV assertions using JavaScript constructs and data types. The key design and implementation principles of these two serialisations of PROV are in compliance with the goals of this study, i.e. understandable and interoperable, hence are a natural choice to support the design of an adaptable provenance profile. For completeness we also explored serializing the provenance graph as PROV-XML <span class="citation">(Hua et al. 2013)</span> as well as PROV-O <span class="citation">(Lebo et al. 2013)</span>, which provides a mapping to Linked Data and ontologies, with potential for rich queries and further integration using a triple store. One challenge here is the wide variety of OWL and RDF formats, we opted for Turtle, N-Triples and JSON-LD, but concluded that requiring all of these PROV and RDF serializations would be an unnecessary burden for other implementations of <em>CWLProv</em>.</p>
<h2 id="sec:cwlprovRO"><em>CWLProv</em> Research Object</h2>
<p>The provenance framework defined in previous section can be satisfied by using a structured approach to share the identified resources. In this section, we define the representation of data and metadata to be shared for a given workflow enactment, stored as multiple files in their native formats. The folder structure of the <em>CWLProv</em> Research Object complies with the <em>BagIt</em> <span class="citation">(Network Working Group 2017)</span> format such that its content and completeness can be verified with any BagIt tool or library (see box <strong>What is BagIt?</strong>). The files used and generated by the workflow are here considered the <em>data payload</em>; the remaining directories include <em>metadata</em> of how the workflow results were created. We systematized the aggregated resources into various collections for better understanding and accessibility for a CWL workflow execution (Figure [fig:RO-format]).</p>
<p><img src="images/RO-structure-NEW-file" alt="image" /></p>
<h3 id="data">data/</h3>
<p><code>data/</code> is the <em>payload</em> collection of the Research Object, in <em>CWLProv</em> this contains all input and output files used in a given workflow enactment. Data files should be labelled and identified based on a hashed checksum rather than derived from its file path during workflow execution. This use of <em>content-addressable</em> reference and storage <span class="citation">(EMC Education Services 2012)</span> simplifies identifier generation for data and helps to avoid local dependencies, e.g. hard-coded file names. However, the workflow execution engine might use other unique identifiers for file objects. It is advised to re-use such identifiers to avoid redundancy and to comply with the system/platform used to run the workflow.</p>
<h3 id="workflow"> workflow/</h3>
<p><em>CWLProv</em> ROs must include a system-independent executable version of the workflow under the <code>workflow/</code> folder. When using CWL, this sub-folder must contain the complete executable <em>workflow specification</em> file, an <em>input file object</em> with parameter settings used to enact the workflow and an <em>output file object</em> generated as a result of workflow enactment. The latter contain details of the workflow outputs such as data files produced by the workflow, but may exclude intermediate outputs.</p>
<p>To ensure RO portability, these file objects may not exactly match the file names at enactment time, as the absolute paths of the inputs are recommended to be replaced with relativized <em>content-addressed</em> paths within the RO, e.g. <code>/home/alice/exp15/sequence.fa</code> is replaced with <code>../data/b1/b1946ac92492d2347c6235b4d2611184</code>. The input file object should also capture any dependencies of the input data files, such as <code>.bam.bai</code> indexes neighbouring <code>.bam</code> (<em>Binary Alignment Map</em>) files. Any folder objects should be expanded to list contained files and their file names at time of enactment.</p>
<p>In the case of a CWL workflow, <em>cwltool</em> can aggregate the CWL description and any referenced external descriptions (such as sub-workflows or command line tool descriptions) into a single workflow file using <code>cwltool --pack</code>. This feature is used in our implementation (details in section <strong></strong>) to rewrite the workflow files, making them re-executable without depending on workflow or commandline descriptions on the file system outside the RO. Other workflow definition approaches, WMS or CWL executors should apply similar features to ensure workflow definitions are executable outside their original file system location.</p>
<p><strong>What is BagIt?</strong></p>
<p><strong>BagIt</strong> is an IETF Internet Standard (RFC8493)<span class="citation">(Network Working Group 2017)</span> that defines a structured file hierarchy for the purpose of digital preservation of data files. BagIt was initiated by the US Library of Congress and the California Digital Library, and is now used by libraries and archives to ensure safe transmission and storage of datasets using “bags”.</p>
<p>A <strong>bag</strong> is indicated by the presence of <code>bagit.txt</code> and a <em>payload</em> of digital content stored as files and sub-folders in the <code>data/</code> folder. Other files are considered <em>tag files</em> to further describe the payload. All the payload files are listed in a <em>manifest</em> with checksums of their byte content, e.g. <code>manifest-sha256.txt</code> and equivalent for tag files in <code>tagmanifest-sha256.txt</code>. Basic metadata can be provided in <code>bag-info.txt</code> as key-value pairs.</p>
<p>A bag can be checked to be <em>complete</em> if all the files listed in the manifests exist, and is also considered <em>valid</em> if the manifest matches the checksum of each file, ensuring they have been correctly transferred.</p>
<p><strong>BDBag</strong> (Big Data bag)<span class="citation">(Chard et al. 2016)</span> is a profile of BagIt that adds a <em>Research Object</em><span class="citation">(Soiland-Reyes, Gamble, and Haines 2014)</span> <code>metadata/manifest.json</code> in JSON-LD <span class="citation">(Sporny et al. 2014)</span> format to contain richer Linked Data annotations that may not fit well in <code>bag-info.txt</code>, e.g. authors of an individual file. BDBags can include a <code>fetch.txt</code> to reference external resources using <em>ARK MinIDs</em> or HTTP URLs, allowing bags that contain large files without necessarily transferring their bytes.</p>
<h3 id="snapshot">snapshot/</h3>
<p><code>snapshot/</code> comprises copies of the workflow and tool specifications files “as-is” at enactment time, without any rewrites, packing or relativizing as described above.</p>
<p>It is recommended to use snapshot resources only for validity checking results and for understanding the workflow enactment, since these files might contain absolute paths or be host-specific, and thus may not be possible to re-enact elsewhere. Preserving these files untouched may nevertheless retain information that could otherwise get lost, e.g. commented out workflow code, or identifiers baked into file names.</p>
<p>A challenge in capturing snapshot files is that they typically live within a file system hierarchy which can difficult to replicate accurately, and may have internal references to other files. In our implementation we utilize <code>cwltool --print-deps</code> to find indirectly referenced files and store their snapshots in a flat folder.</p>
<h3 id="metadata">metadata/</h3>
<p>Each <em>CWLProv</em> RO must contain an RO manifest file <code>metadata/manifest.json</code> and two sub-directories <code>metadata/logs</code> and <code>metadata/provenance</code>. The RO manifest, part of the BDBag <span class="citation">(Chard et al. 2016)</span> profile, follows the JSON-LD structure defined for Research Object Bundles <span class="citation">(Soiland-Reyes, Gamble, and Haines 2014)</span> and can provide structured Linked Data for each file in the RO, like file type and creation date. Further detail about the manifest file contents is documented on GitHub as <em>CWLProv</em> specification <span class="citation">(Soiland-Reyes, Khan, and Crusoe 2018)</span>.</p>
<p>Any raw log information from the workflow enactment should be made available in <code>metadata/logs</code>. This typically includes the actual commands executed for each step. Similar to the snapshot files, log files may however be difficult to process outside the original enactment system. An example of such processing is <em>CWL-metrics</em> <span class="citation">(Ohta, Tanjo, and Ogasawara 2019)</span>, which post-process cwltool log files to capture runtime metrics of individual Docker containers.</p>
<p>Capturing the details of a workflow execution require rich metadata in provenance files (see section <strong></strong>). These should exist in the sub-folder <code>metadata/provenance</code>. It is recommended to make the availability of a <em>primary</em> provenance file mandatory, which should conform with the PROV-N <span class="citation">(Moreau, Missier, Cheney, et al. 2013)</span> format. This file describes the top-level workflow execution. As described in <em>Level 2</em> (Section <strong></strong>), it is quite possible to have nested workflows. In that case, a separate provenance file for each nested workflow execution should be included in this folder. If there are additional formats of provenance files such as PROV-JSON <span class="citation">(Huynh et al. 2013)</span>, PROV-XML <span class="citation">(Hua et al. 2013)</span>, PROV-O <span class="citation">(Lebo et al. 2013)</span> etc, then these should be included in the said folder with a declaration using <code>conformsTo</code> to declare their formats in the RO manifest being mandatory. The nested workflow profile should be named such that there is a link between the respective step in the primary workflow and the nested workflow preferably using unique identifiers.</p>
<p>As the PROV-DM has a generalized structure, there might be some provenance aspects specific to particular workflows that are hard to capture if only using PROV-N, hence ontologies such as <em>wfdesc</em> <span class="citation">(Soiland-Reyes et al. 2016b)</span> can be used to describe the abstract representation of the workflow and its steps. Use of <em>wfprov</em> <span class="citation">(Soiland-Reyes et al. 2016a)</span> to capture some workflow provenance aspects is also encouraged. Alternative extensions such as ProvOne <span class="citation">(Cao et al. 2016)</span> can also be utilized if the WMS or workflow executor is using these extensions already.</p>
<p><em>CWLProv</em> reuses Linked Data standards like JSON-LD <span class="citation">(Sporny et al. 2014)</span>, W3C PROV <span class="citation">(Moreau, Missier, Belhajjame, et al. 2013)</span> and Research Object <span class="citation">(Hettne et al. 2014)</span>. A challenge with Linked Data in distributed and desktop computing is how to make identifiers that are absolute URIs and hence globally unique. For example, for <em>CWLProv</em> a workflow may be executed by an engine that does not know where its workflow provenance will be stored, published or finally integrated. To this end <em>CWLProv</em> generators should use the proposed <em>arcp</em> <span class="citation">(Soiland-Reyes and Cáceres 2018a)</span> URI scheme to map local file paths within the RO BagIt folder structure to absolute URIs for use within the RO manifest and associated PROV traces. Consumers of <em>CWLProv</em> ROs that do not contain an arcp-based External-Identifier should generate a temporary arcp base to safely resolve any relative URI references not present in the <em>CWLProv</em> folder. Implementations processing a <em>CWLProv</em> RO may convert arcp URIs to local <code>file:///</code> or <code>http://</code> URIs depending on how and where the <em>CWLProv</em> RO was saved, e.g. using the “arcp.py” library <span class="citation">(Soiland-Reyes and Cáceres 2018b)</span>.</p>
<h2 id="sec:provenanceprofile">Retrospective Provenance Profile</h2>
<p><span>l l l l L</span> PROV type &amp; Subtype &amp; Relation &amp; Range &amp; Recommendation<br />
<strong>Plan</strong> &amp; wfdesc:Workflow &amp; wfdesc:hasSubProcess &amp; wfdesc:Process &amp; R9-workflow<br />
&amp; wfdesc:Process &amp; &amp;<br />
<strong>Activity</strong> &amp; wfprov:WorkflowRun &amp; wasAssociatedWith &amp; wfprov:WorkflowEngine &amp; R8-environment<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>hadPlan </span> &amp;   wfdesc:Workflow &amp; R9-workflow, R17-executable<br />
&amp; &amp; wasStartedBy &amp; wfprov:WorkflowEngine &amp; R8-environment<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>atTime </span> &amp;   <span> <em>ISO8601 timestamp</em> </span> &amp; R13-provenance<br />
&amp; &amp; wasStartedBy &amp; wfprov:WorkflowRun &amp; R9-workflow<br />
&amp; &amp; wasEndedBy &amp; wfprov:WorkflowEngine &amp; R8-environment<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>atTime </span> &amp;   <span> <em>ISO8601 timestamp</em> </span> &amp; R13-provenance<br />
&amp; wfprov:ProcessRun &amp; wasStartedBy &amp; wfprov:WorkflowRun &amp; R10-software<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>atTime </span> &amp;   <span> <em>ISO8601 timestamp</em> </span> &amp; R14-provenance<br />
&amp; &amp; used &amp; wfprov:Artifact &amp; R11-raw-data<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>role </span> &amp;   wfdesc:InputParameter &amp; R1-parameters<br />
&amp; &amp; wasAssociatedWith &amp; wfprov:WorkflowRun &amp; R9-workflow<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>hadPlan </span> &amp;   wfdesc:Process &amp; R17-executable, R16-format<br />
&amp; &amp; wasEndedBy &amp; wfprov:WorkflowRun &amp; R13-provenance<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>atTime </span> &amp;   <span> <em>ISO8601 timestamp</em> </span> &amp; R13-provenance<br />
&amp; SoftwareAgent &amp; wasAssociatedWith &amp; wfprov:ProcessRun &amp; R8-environment<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>cwlprov:image </span> &amp;   <span> <em>docker image id</em> </span> &amp; R4-sw-version<br />
<strong>SoftwareAgent</strong> &amp; wfprov:WorkFlowEngine &amp; wasStartedBy &amp; Person <span> <em>ORCID</em> </span> &amp; R12-attribution<br />
&amp; &amp; label &amp; <span> <em>cwltool <code>--</code>version</em> </span> &amp; R4-sw-version<br />
<strong>Entity</strong> &amp; wfprov:Artefact &amp; wasGeneratedBy &amp; wfprov:Processrun &amp; R3-intermediate, R7-identifier<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>role </span> &amp;   wfdesc:OutputParameter &amp; R1-parameters<br />
<strong>Collection</strong> &amp; wfprov:Artefact &amp; hadMember &amp; wfprov:Artefact &amp; R3-intermediate<br />
&amp; Dictionary &amp; hadDictionaryMember &amp; wfprov:Artefact &amp;<br />
&amp; &amp; <span>   <span><span class="math inline">↳</span>  </span>pairKey </span> &amp;   <span> <em>filename</em> </span> &amp; R7-identifier<br />
</p>
<p>Indentation with <span><span class="math inline">↳</span>  </span>indicates n-ary relationships which are expressed differently depending on PROV syntax. Namespaces: <a href="http://www.w3.org/ns/prov#" class="uri">http://www.w3.org/ns/prov#</a> (default), <a href="http://purl.org/wf4ever/wfdesc#" class="uri">http://purl.org/wf4ever/wfdesc#</a> (<em>wfdesc</em>), <a href="http://purl.org/wf4ever/wfprov#" class="uri">http://purl.org/wf4ever/wfprov#</a> (<em>wfprov</em>), <a href="https://w3id.org/cwl/prov#" class="uri">https://w3id.org/cwl/prov#</a> <em>cwlprov</em>)</p>
<p>As stated earlier, the primary provenance file should conform to the PROV-N <span class="citation">(Moreau, Missier, Cheney, et al. 2013)</span> serialisation of PROV data model, and may optionally use ontologies specific to the workflow execution. The key features used in the structure of the retrospective provenance profile for a CWL workflow enactment in <em>CWLProv</em> are listed in Table [tab:provProfile]). These features are not tied to any platform or workflow definition approach and hence can be used to document retrospective provenance of any workflow irrespective of the workflow definition approach.</p>
<p>The core mapping is following the PROV data model as in Figure [fig:prov-dm]): The PROV <em>Activity</em> represent the duration of a workflow run, as well as individual step executions, which <em>used</em> file and data (<em>Entity</em>), which again may be <em>wasGeneratedBy</em> previous step activities. The workflow engine (e.g. cwltool) is the <em>Agent</em> controlling these activities according to the workflow definition (<em>Plan</em>).</p>
<p>PROV is a general standard not specific to workflows, and lacks features to relate a <em>plan</em> (i.e. a workflow description) with sub-plans and workflow-centric retrospective provenance elements e.g. specific workflow enactment and its related steps enactment. We have utilized <em>wfdesc</em> and <em>wfprov</em> to represent few elements of prospective and retrospective provenance respectively. In addition, the provenance profile documented details of all the uniquely identified <em>activities</em> e.g. workflow enactment and related command line tool invocations, their associated <em>entities</em> (e.g. input and output data artefacts, input configuration files, workflows and command line tool specifications). The profile also documents the relationship between activities such as which activity (workflow enactment) was responsible for starting and ending another activity (command line tool invocation).</p>
<p>As described in Section <strong></strong>, in order to achieve maximum <em>white-box</em> provenance, the inner workings of a nested workflow should also be included in the provenance trace. If a step represents a nested workflow, a separate provenance profile is included in the RO. Moreover, in the parent workflow trace, this relationship is recorded using <em>has_provenance</em> as an attribute of the <em>Activity</em> step which refers to the profile of the nested workflow.</p>
<h1 id="sec:demo">Practical Realisation of <em>CWLProv</em></h1>
<p><em>CWLProv</em> <span class="citation">(Soiland-Reyes, Khan, and Crusoe 2018)</span> provides a format that can be adopted by any workflow executor or platform, provided that the underlying workflow definition approach is at least as declarative as CWL, i.e. it captures the necessary components described in Section <strong></strong>. In the case of CWL, as long as the conceptual constructs are common amongst the available implementations and executors, a workflow enactment can be represented in <em>CWLProv</em> format. To demonstrate the practical realisation of the proposed model we consider a Python-based reference implementation of CWL <em>cwltool</em>.</p>
<p><em>cwltool</em> is a feature complete reference implementation of CWL. It provides extensive validation of CWL files as well as offering a comprehensive set of test cases to validate new modules introduced as extensions to the existing implementation. Thus it provides the ideal choice for implementing <em>CWLProv</em> for provenance support and resource aggregation. The existing classes and methods of the implementation were utilized to achieve various tasks such as packaging of the workflow and all associated tool specifications together. In addition, the existing python library <em>prov</em> <span class="citation">(“Prov 1.5.2” 2018)</span> was used to create a provenance document instance and populate it with the required artefacts generated as the workflow enactment proceeds.</p>
<p>It should be noted that we elected to implement <em>CWLProv</em> in the reference implementation <em>cwltool</em> instead of the more scalable and production-friendly CWL implementations like Toil <span class="citation">(Vivian et al. 2017)</span>, Arvados <span class="citation">(“Arvados - Open Source Big Data Processing and Bioinformatics” 8AD–11AD)</span>, Rabix <span class="citation">(Kaushik et al. 2017)</span>, CWL-Airflow <span class="citation">(Kotliar, Kartashov, and Barski 2018)</span> or Cromwell <span class="citation">(Voss, Auwera, and Gentry 2017)</span>. An updated list of implementations is available at the CWL homepage.<a href="#fn2" class="footnoteRef" id="fnref2"><sup>2</sup></a> Compared to <em>cwltool</em> these generally have extensive scheduler and cloud compute support, and extensions for large data transfer and storage, and should therefore be considered for any adopters of the Common Workflow Language. In this study we have however focused on <em>cwltool</em> as its code base was found to be easy to adapt for rich provenance capture without having to modify subsystems for distributed execution or data management, and as a reference implementation better informing us on how to model <em>CWLProv</em> for the general case rather than being tied into execution details of the more sophisticated CWL workflow engines.</p>
<p><em>CWLProv</em> support for <em>cwltool</em> is built as an optional module which when invoked as <em>“cwltool <code>--</code>provenance ro/ workflow.cwl job.json”</em>, will automatically generate an RO with the given folder name <em>ro/</em> without requiring any additional information from the user. Each input file is assigned a hash value and placed in the folder <em>ro/data</em>, making it content-addressable to avoid local dependencies (Figure [fig:processflow]).</p>
<p>[t!]</p>
<p>In order to avoid including information about attribution without consent of the user, we introduce an additional flag <em>“ <code>--</code>enable-user-provenance”</em>. If a user provides the options <em><code>--</code>orcid</em> and <em><code>--</code>full-name</em>, this information will be included in the provenance profile related to user attribution. Enabling <em>“ <code>--</code>enable-user-provenance”</em> and not providing the full name or ORCID will store user account details from the local machine for attribution, i.e. the details of the <em>agent</em> that enacted the workflow.</p>
<p>The workflow and command line tool specifications are aggregated in one file to create an executable workflow and placed in folder <em>ro/workflow</em>. This folder also contains transformed input job objects containing the input parameters with references to artefacts in the <em>ro/data</em> based on relativising the paths present in the input object. These two files are sufficient to re-enact the workflow, provided the other required artefacts are also included in the RO and comply to the <em>CWLProv</em> format. The <em>cwltool</em> control flow <span class="citation">(“Common-Workflow-Language/Cwltool:Common Workflow Language Reference Implementation” 2016)</span> indicates the points when the execution of the workflow and command line tools involved in the workflow enactment start, end and how the output is reported back. This information and the artefacts are captured and stored in the RO.</p>
<p>When the execution of a workflow begins, <em>CWLProv</em> extensions to <em>cwltool</em> generate a provenance document (using the <em>prov</em> library) which includes default namespaces for the workflow enactment <em>“activity”</em>. The attribution details as an <em>agent</em> are also added at this stage if user provenance capture is enabled, e.g. to answer “who ran the workflow?”. Each step of the workflow can correspond to either a command line tool or another nested workflow referred to as a <em>sub-workflow</em> in the CWL documentation. For each nested workflow, a separate provenance profile is initialized recursively to achieve a <em>white-box</em> finer-grained provenance view as explained in Section <strong></strong>. This profile is continually updated throughout the nested workflow enactment. Each step is identified by a unique identifier and recorded as an <em>activity</em> in the parent workflow provenance profile, i.e. the <em>“primary profile”</em>. The <em>nested</em> workflow is recorded as a step in the <em>primary profile</em> using the same identifier as the “nested workflow enactment activity” identifier in the respective provenance profile. For each step in the activity, the start time and association with the workflow activity is created and stored as part of the overall provenance to answer the question “when did it happen?”.</p>
<p>The data used as input by these steps is either provided by the user or produced as an intermediate result from the previous steps. In both cases, the <em>Usage</em> is recorded in the respective provenance profile using checksums as identifiers to answer the question “what was used?”. The non-file input parameters such as strings and integers are stored “as-is” using an additional optional argument, <em>prov:value</em>. Upon completion, each step typically generates some data. The provenance profile records the generation of outputs at the step level to record “what was produced?” and “which process produced it?”. Once all steps complete, the workflow outputs are collected and the generation of these outputs at the workflow level are recorded in the provenance profile. Moreover, using the checksum of these files generated by the <em>cwltool</em>, content-addressable copies are saved in the folder <em>ro/data</em>. The provenance profile refers to these files using the same checksum such that they are traceable or can be used for further analysis if required. The workflow specification, command line tool specifications and JSON job file is archived in the <em>ro/snapshot</em> folder to preserve the actual workflow history.</p>
<p>This prototype implementation provides a model and guidance for workflow platforms and executors to identify their respective features that can be utilized in devising their own implementation of <em>CWLProv</em>.</p>
<h2 id="achieving-recommendations-with-provenance-levels">Achieving recommendations with provenance levels</h2>
<p>Table [tab:fulfilling] map the best practices and recommendations from Table [tab:recommendation:wide] to the Levels of Provenance (Figure [fig:levels]). The shown methods and implementation readiness indicate to which extent the recommendations are addressed by the implementation of <em>CWLProv</em> (detailed in this section).</p>
<p>Note that other approaches may solve this mapping differently. For instance, Nextflow <span class="citation">(Di Tommaso et al. 2017)</span> may fulfill <em>R18-resource-use</em> at Provenance as it can produce trace reports with hardware resource usage per task execution <span class="citation">(Centre for Genomic Regulation (CRG) 2018)</span>, but not for the overall workflow. While a Nextflow trace report is a separate CSV file with implementation-specific columns, our planned <em>R18-resource-use</em> approach for CWL is to combine <em>CWL-metrics</em> <span class="citation">(Ohta, Tanjo, and Ogasawara 2018)</span>, permalinks and the standard <em>GFD.204</em> <span class="citation">(Cristofori et al. 2013)</span> to further relate resource use with and provenance within the <em>CWLProv</em> Research Object.</p>
<p>In addition to following the recommendations from Table [tab:recommendation:wide] through computational methods, the workflow authors are also required to exercise <em>best practices for workflow design and authoring</em>. For instance, to achieve <em>R1-parameters</em> the workflow must be written in such a way that parameters are exposed and documented at workflow level, rather than hard-coded within an underlying Python script. Similarly, while the CWL format support rich details of user annotations that can fulfill <em>R6-annotation</em>, for these to survive into a Research Object at execution time, such annotation capabilities must actually be used by workflow authors instead of unstructured text files.</p>
<p>It should be a goal of a scientific WMS to guide users towards achieving the required level of the provenance framework through automation where possible. For instance a user may in the workflow have specified a Docker container image without preserving the version, but the provenance log could still record the specific container version used at execution time, achieving <em>R4-sw-version</em> retrospectively by computation rather than relying on a prospective declaration in the workflow definition.</p>
<table>
<caption>Recommendations and provenance levels implemented in <em>CWLProv</em><span data-label="tab:fulfilling"></span></caption>
<thead>
<tr class="header">
<th align="left">Recommendation</th>
<th align="center">L0</th>
<th align="center">L1</th>
<th align="center">L2</th>
<th align="center">L3</th>
<th align="left">Methods</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">R1-parameters</td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="left">CWL, BP</td>
</tr>
<tr class="even">
<td align="left">R2-automate</td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">CWL, Docker</td>
</tr>
<tr class="odd">
<td align="left">R3-intermediate</td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">PROV, RO</td>
</tr>
<tr class="even">
<td align="left">R4-sw-version</td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="left">CWL, Docker, PROV</td>
</tr>
<tr class="odd">
<td align="left">R5-data-version</td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="left">CWL, BP</td>
</tr>
<tr class="even">
<td align="left">R6-annotation</td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"><span class="math inline">$\coasterisk$</span></td>
<td align="left">CWL, RO, BP</td>
</tr>
<tr class="odd">
<td align="left">R7-described</td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">CWL, RO</td>
</tr>
<tr class="even">
<td align="left">R7-identifier</td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"><span class="math inline">•</span></td>
<td align="left">RO, CWLProv</td>
</tr>
<tr class="odd">
<td align="left">R8-environment</td>
<td align="center"></td>
<td align="center"><span class="math inline">$\coasterisk$</span></td>
<td align="center"><span class="math inline">$\coasterisk$</span></td>
<td align="center"></td>
<td align="left">GFD.204</td>
</tr>
<tr class="even">
<td align="left">R9-workflow</td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="left">CWL, wfdesc</td>
</tr>
<tr class="odd">
<td align="left">R10-software</td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="left">CWL, Docker</td>
</tr>
<tr class="even">
<td align="left">R11-raw-data</td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">CWLProv, BP</td>
</tr>
<tr class="odd">
<td align="left">R12-attribution</td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">RO, CWL, BP</td>
</tr>
<tr class="even">
<td align="left">R13-provenance</td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="left">PROV, RO</td>
</tr>
<tr class="odd">
<td align="left">R14-diagram</td>
<td align="center"><span class="math inline">∘</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="center"><span class="math inline">$\coasterisk$</span></td>
<td align="left">CWL, RO</td>
</tr>
<tr class="even">
<td align="left">R15-open-source</td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">CWL, BP</td>
</tr>
<tr class="odd">
<td align="left">R16-format</td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"><span class="math inline">•</span></td>
<td align="left">CWL, BP</td>
</tr>
<tr class="even">
<td align="left">R17-executable</td>
<td align="center"><span class="math inline">∘</span></td>
<td align="center"><span class="math inline">•</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">CWL, Docker</td>
</tr>
<tr class="odd">
<td align="left">R18-resource-use</td>
<td align="center"></td>
<td align="center"><span class="math inline">$\coasterisk$</span></td>
<td align="center"><span class="math inline">$\coasterisk$</span></td>
<td align="center"></td>
<td align="left">CWL, GFD.204</td>
</tr>
<tr class="even">
<td align="left">R19-example</td>
<td align="center"><span class="math inline">$\coasterisk$</span></td>
<td align="center"><span class="math inline">∘</span></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">RO, BP</td>
</tr>
</tbody>
</table>
<p><strong>CWL</strong>: Common Workflow Language and embedded annotations</p>
<p><strong>RO</strong>: Research Object model and BagIt</p>
<p><strong>PROV</strong>: W3C Provenance model</p>
<p><strong>CWLProv</strong>: Additional attributes in PROV</p>
<p><strong>wfdesc</strong>: Prospective provenance in PROV</p>
<p><strong>BP</strong>: Best Practice need to be followed manually</p>
<p><span class="math inline">•</span> Implemented</p>
<p><span class="math inline">∘</span> Partially implemented</p>
<p><span class="math inline">$\coasterisk$</span> Implementation planned/ongoing</p>
<h1 id="sec:evaluation">CWLProv Evaluation with Bioinformatics Workflows</h1>
<p><em>CWLProv</em> as a standard supports <em>syntactic</em>, <em>semantic</em> and <em>pragmatic</em> interoperability (defined in Section <strong></strong>) of a given workflow and its associated results. We have defined a <em>“common data format”</em> for workflow sharing and publication such that any executor or WMS with CWL support can interpret this information and make use of it. This ensures the <em>syntactic</em> interoperability between the workflow executors on different computing platforms. Similarly the <em>“content”</em> of the shared aggregation artefact as a workflow-centric RO is unambiguously defined, thus ensuring uniform representation of the workflow and its associated results across different platforms and executors hence supporting <em>semantic</em> interoperability. With <em>Level 3</em> provenance satisfied providing domain-specific information along with level 0-2 provenance tracking, we posit that <em>CWLProv</em> would be able to accomplish <em>pragmatic</em> interoperability by providing unambiguous information about the <em>“context”</em>, <em>“application”</em> and <em>“use”</em> of the shared/published workflow-centric ROs. Hence, extension of the current implementation (described in section [sec:demo]) in future to include domain-rich information in the provenance traces and the <em>CWLProv RO</em> will result in pragmatic interoperability.</p>
<p>To demonstrate the interoperability and portability of the proposed solution, we evaluate <em>CWLProv</em> and its reference implementation using open source bioinformatics workflows available on GitHub from different research initiatives and from different developers. Conceptually, these workflows are selected for evaluation due to their excessive use in real-life data analyses and variety of the input data. Alignment workflow is included in the evaluation as it is one of the most time consuming yet mandatory steps in any variant calling workflow. Practically, choosing the workflows by these particular groups out of numerous existing implementations is justified in each section below.</p>
<h2 id="sec:rnaseq-wf">RNA-seq Analysis Workflow</h2>
<p>[t!] <img src="images/rnaseq-cwlviewer-half.png" alt="image" /></p>
<p>RNA sequencing (RNA-seq) data generated by Next Generation Sequencing (NGS) platforms is comprised of short sequence reads that can be aligned to a reference genome, where the alignment results form the basis of various analyses such as quantitating transcript expression; identifying novel splice junctions and isoforms and differential gene expression <span class="citation">(Dobin and Gingeras 2015)</span>. RNA-seq experiments can link phenotype to gene expression and are widely applied in multi-centric cancer studies <span class="citation">(Cohen-Boulakia et al. 2017)</span>. Computational analysis of RNA-seq data is performed by different techniques depending on the research goals and the organism under study <span class="citation">(Conesa et al. 2016)</span>. The workflow <span class="citation">(“TopMed Rna-Seq Workflow” 2018)</span> included in this case study has been defined in CWL by one of the teams <span class="citation">(“Heliumdatacommons” 2017)</span> participating in NIH Data Commons initiative <span class="citation">(“Data Commons | Nih Common Fund” 2018)</span>, a large research infrastructure program aiming to make digital objects (such as data generated during biomedical research and software/tools required to utilize such data) shareable and accessible and hence aligned with the FAIR principles <span class="citation">(Wilkinson et al. 2016)</span>.</p>
<p>This workflow (Figure [fig:rna-seq]), designed for the pilot phase of the NIH Data Commons initiative <span class="citation">(“Heliumdatacommons/Cwl_workflows: Example Cwl Workflows That Run on Team Helium Pivot Architecture.” 11AD–10AD)</span>, adapts the approach and parameter settings of Trans-Omics for precision Medicine (TOPMed) <span class="citation">(“Trans-Omics for Precision Medicine (Topmed) Program | National Heart, Lung, and Blood Institute (Nhlbi)” 2014)</span>. The RNA-seq pipeline originated from the Broad Institute <span class="citation">(“Gtex Rna-Seq Pipeline” 2017)</span>. There are in total five steps in the workflow starting from: 1) Read alignment using STAR <span class="citation">(Dobin et al. 2012)</span> which produces aligned BAM files including the Genome BAM and Transcriptome BAM. 2) The Genome BAM file is processed using Picard MarkDuplicates <span class="citation">(“Tool Documentation: MarkDuplicates”)</span> producing an updated BAM file containing information on duplicate reads (such reads can indicate biased interpretation). 3) SAMtools index <span class="citation">(Li et al. 2009)</span> is then employed to generate an index for the BAM file, in preparation for the next step. 4) The indexed BAM file is processed further with RNA-SeQC <span class="citation">(DeLuca et al. 2012)</span> which takes the BAM file, human genome reference sequence and Gene Transfer Format (GTF) file as inputs to generate transcriptome-level expression quantifications and standard quality control metrics. 5) In parallel with transcript quantification, isoform expression levels are quantified by RSEM <span class="citation">(Li and Dewey 2011)</span>. This step depends only on the output of the STAR tool, and additional RSEM reference sequences.</p>
<p>For testing and analysis, the workflow author provided example data created by down-sampling the read files of a TOPMed public access data <span class="citation">(Seo et al. 2012)</span>. Chromosome 12 was extracted from the <em>Homo Sapien Assembly 38</em> reference sequence and provided by the workflow authors. The required GTF and RSEM reference data files are also provided. The workflow is well-documented with a detailed set of instructions of the steps performed to down-sample the data are also provided for transparency. The availability of example input data, use of containerization for underlying software and detailed documentation are important factors in choosing this specific CWL workflow for <em>CWLProv</em> evaluation.</p>
<h2 id="sec:align">Alignment Workflow</h2>
<p>Alignment is an essential step in variant discovery workflows and considered an obligatory <em>pre-processing</em> stage according to Best Practices by the Broad Institute <span class="citation">(“GATK | Bp Doc #11165 | Data Pre-Processing for Variant Discovery” 2018)</span>. The purpose of this stage is to filter low-quality reads before variant calling or other interpretative steps <span class="citation">(Xu 2018)</span>. The workflow for alignment is designed to operate on raw sequence data to produce analysis-ready BAM files as the final output. The typical steps followed include file format conversions, aligning the read files to the reference genome sequence, and sorting the resulting files.</p>
<p>[b!] <img src="images/new_alignment.png" alt="image" style="width:95.0%" /></p>
<p>The CWL alignment workflow <span class="citation">(“Topmed-Workflows/Topmed-Alignment.cwl at Cwlprov_testing · Farahzkhan/Topmed-Workflows”)</span> included in this evaluation (Figure [fig:align]) is designed by Data Biosphere <span class="citation">(“Data Biosphere” 2018)</span>. It adapts the alignment pipeline <span class="citation">(“Statgen/Docker-Alignment: Dockerfile for Alignment” 2017)</span> originally developed at Abecasis Lab, The University of Michigan <span class="citation">(“Abecasis Lab - Genome Analysis Wiki” 2017)</span>. This workflow is also part of NIH Data Commons initiative (as ) and comprises of four stages. First step, “Pre-align” accepts a Compressed Alignment Map (CRAM) file (a compressed format for BAM files developed by European Bioinformatics Institute (EBI) <span class="citation">(Cochrane et al. 2012)</span>) and human genome reference sequence as input and using underlying software utilities of SAMtools such as view, sort and fixmate returns a list of fastq files which can be used as input for the next step. The next step “Align” also accepts the human reference genome as input along with the output files from “Pre-align” and uses BWA-mem <span class="citation">(Li 2013)</span> to generate aligned reads as BAM files. SAMBLASTER <span class="citation">(Faust and Hall 2014)</span> is used to mark duplicate reads and SAMtools view to convert read files from SAM to BAM format. The BAM files generated after “Align” are sorted with “SAMtool sort”. Finally these sorted alignment files are merged to produce single sorted BAM file using SAMtools merge in “Post-align” step. The authors provide an example CRAM file, <em>Homo Sapien Assembly 38</em> reference genome along with its index files to be used as inputs for testing and analysis of the workflow.</p>
<h2 id="somatic-variant-calling-workflow">Somatic Variant Calling Workflow</h2>
<p>Variant discovery analysis for high-throughput sequencing data is a widely used bioinformatics technique, focused on finding genetic associations with diseases, identifying somatic mutations in cancer and characterizing heterogeneous cell populations <span class="citation">(“Introduction to Variant Calling” 2014)</span>. The <em>pre-processing</em> explained for the Alignment workflow is part of any variant calling workflow as reads are classified and ordered as part of the variant discovery process. Numerous variant calling algorithms have been developed depending on the input data characteristics and the specific application area <span class="citation">(Xu 2018)</span>. Somatic variant calling workflows are designed to identify somatic (non-inherited) variants in a sample - generally a cancer sample - by comparing the set of variants present in a sequenced tumour genome to a non-tumour genome from the same host <span class="citation">(Saunders et al. 2012)</span>. The set of tumour variants is a super-set of the set of host variants, and somatic mutations can be identified through various algorithmic approaches to subtracting host familial variants. Each somatic variant calling workflow typically consists of three stages: pre-processing; variant evaluation and post-filtering.</p>
<p>The somatic variant calling workflow (Figure [fig:somatic]) included in this case study is designed by Blue Collar Bioinformatics (bcbio) <span class="citation">(“Blue Collar Bioinformatics”)</span>, a community-driven initiative to develop best-practice pipelines for variant calling, RNA-seq and small RNA analysis workflows. According to the documentation, the goal of this project is to facilitate the automated analysis of high throughput data by making the resources <em>quantifiable</em>, <em>analyzable</em>, <em>scalable</em>, <em>accessible</em> and <em>reproducible</em>. All the underlying tools are containerized facilitating software use in the workflow. The somatic variant calling workflow defined in CWL is available on GitHub <span class="citation">(“Somatic Variant Calling Workflow” 2018)</span> and equipped with a well defined test dataset.</p>
<p>[t!] <img src="images/2variantcalling.png" alt="image" height="453" /></p>
<h2 id="eval-activity">Evaluation Activity</h2>
<p>This section describes the evaluation of cross-executor and cross-platform interoperability of <em>CWLProv</em>. To test cross-executor interoperability, two CWL executors <em>cwltool</em> and <em>toil-cwl-runner</em> were selected. <em>toil-cwl-runner</em> is an open source Python workflow engine supporting robust cross-platform workflow execution on Cloud and High Performance Computing (HPC) environments <span class="citation">(Vivian et al. 2017)</span>. The two operating system platforms utilized in this analysis were MacOS and Ubuntu Linux. For the Linux OS, a 16-core Linux instance with 64GB RAM was launched on the Australian National eResearch Collaboration Tools and Resources (NeCTAR) research cloud <span class="citation">(“Nectar Cloud - Nectar”)</span>. To cater for the storage requirements, a 1000GB persistent volume was attached to this instance. For MacOS, a local system with 16GB RAM, 250GB storage and 2.8 GHz Intel Core i7 processor was used. These platforms were selected to cater for the required storage and compute resources of the workflows described above. The reference genome provided with was not down-sampled and hence this workflow required most resources among the three evaluated.</p>
<p>It is worth mentioning that this evaluation does not include details of the installation process for <em>cwltool</em>, <em>toil-cwl-runner</em> and <em>Docker</em> on systems described above. To create <em>CWLProv</em> ROs during workflow execution, it is necessary to use the CWL reference runner (<em>cwltool</em>) until this practice spreads to other CWL implementations. Moreover, it is assumed that the software container (Docker) should also be installed on the system to use the workflow definitions aggregated in a given <em>CWLProv</em> RO.</p>
<p>In addition, the resource requirements (identified in <em>R18-resource-use</em> and discussed in Section <strong></strong>) should also be satisfied by choosing a system with enough compute and storage resources for successful enactment. The systems used in this case study should be a reference when selecting a system as inadequate compute and storage resources such as insufficient RAM or number of cores will hinder the successful re-enactment of workflows using these ROs. The hardware requirements may also vary if a different dataset is used as input to re-enact the workflow using the methods aggregated in the RO. In that case, the end user must ensure availability of adequate compute and storage resources by choosing a system that meets the required specifications <span class="citation">(Kanwal, Lonie, and Sinnott 2017)</span>.</p>
<p>Since the <em>CWLProv</em> implementation is demonstrated for one of the executors (<em>cwltool</em>), currently a <em>CWLProv</em> RO for any workflow can only be produced using <em>cwltool</em>. Hence, in this activity the workflows are initially enacted using just <em>cwltool</em> (Table [tab:eval]). The outline of the steps performed to analyse <em>CWLProv</em> for each case study is as follows.</p>
<ol>
<li><p>The workflow was enacted using <em>cwltool</em> to produce a RO on a MacOS computer.</p>
<ol>
<li><p>The resulting RO and aggregated resources were used to re-enact the workflow using <em>toil-cwl-runner</em> on the same MacOS computer;</p></li>
<li><p>The RO produced in step I was transferred to the cloud-based Linux instance used in this activity;</p></li>
<li><p>On the cloud-based Linux environment and only utilizing the resources aggregated in the RO, the workflow was re-enacted using <em>cwltool</em> and <em>toil-cwl-runner</em>.</p></li>
</ol></li>
<li><p>The workflow was enacted using <em>cwltool</em> to produce a RO on Linux.</p>
<ol>
<li><p>The resulting RO and aggregated resource were utilized to re-enact the workflow using <em>toil-cwl-runner</em> on the same cloud-based Linux instance;</p></li>
<li><p>The RO produced in step II was transferred to the MacOS computer used in this activity;</p></li>
<li><p>On the MacOS computer and only utilizing the resources aggregated in the RO, the workflow was re-enacted using <em>cwltool</em> and <em>toil-cwl-runner</em>.</p></li>
</ol></li>
</ol>
<p>The <em>CWLProv</em> ROs produced as a results of this activity are published on Mendeley Data <span class="citation">(Khan and Soiland-Reyes 2018b; Khan and Soiland-Reyes 2018a; Khan and Soiland-Reyes 2018c)</span> with mirrors on Zenodo.</p>
<h2 id="sec:eval-results">Evaluation Results</h2>
<p>The steps described above were taken to produce ROs which were then used to re-enact the workflows (outlined in Table [tab:eval]), without any further changes required. This demonstration illustrated the syntactic and semantic interoperability of the workflows across different systems. It shows that <strong>both CWL executors were able to <em>exchange</em>, <em>comprehend</em> and <em>use</em> the information represented as <em>CWLProv</em> ROs</strong>. The current implementation described in section <strong></strong> does not resolve <em>Level 3</em>. Hence, the inclusion of domain-specific annotations referring to scientific context to address pragmatic interoperability is identified as crucial future direction and further detailed in section <strong></strong>.</p>
<p><span>c c c</span> Enact-produce RO with &amp; Re-enact using RO with &amp; Status<br />
<em>cwltool</em> on MacOS &amp; <em>toil-cwl-runner</em> on MacOS &amp;<br />
&amp; <em>cwltool</em> on Linux &amp;<br />
&amp; <em>toil-cwl-runner</em> on Linux &amp;<br />
<em>cwltool</em> on Linux &amp; <em>toil-cwl-runner</em> on Linux &amp;<br />
&amp; <em>cwltool</em> on MacOS &amp;<br />
&amp; <em>toil-cwl-runner</em> on MacOS &amp;<br />
</p>
<h3 id="cwlprov-and-interoperability"><em>CWLProv</em> and Interoperability</h3>
<p>CWL already builds on technologies such as JavaScript Object Notation for Linked Data (JSON-LD) <span class="citation">(Sporny et al. 2014)</span> for data modeling and Docker <span class="citation">(Merkel 2014)</span> to support portability of the run-time environments. The portability and interoperability as basic principles of the underlying workflow definition approach for any workflow-centric analysis implies that the analysis should also be portable and interoperable. However, the workflow definition/specification alone is insufficient when dealing with commandline tool specifications, data, and input configuration files used in the analysis if these are not readily available.</p>
<p><em>CWLProv</em> ensures availability of these resources for a given analysis conforming to the framework defined in Section <strong></strong>. The input configurations are saved as <em>primary-job.json</em> in folder <em>workflow/</em> and refer to the input data contained in the payload <em>data/</em> folder of the given RO. In this way, availability of data aggregated with the analysis is made possible. Existing features of <em>cwltool</em> are used to generate the CWL workflow specification file containing all of the commandline tool specifications referred to in the workflow specification and placed in the same <em>workflow/</em> folder.</p>
<p>One might argue that copying a folder tree might serve the same purpose but in that case we again will be relying on users to put substantial amount of effort on top of the actual analysis, i.e. they would have to carefully structure their directories to be aligned with the workflow creators. Instead CWL encourages researchers to utilize container technologies such as Docker, Singularity, or software packaging systems like Debian (Med) or Bioconda to ensure availability of underlying tools as recommended by numerous studies <span class="citation">(Belhajjame et al. 2015; Kanwal et al. 2017; Garijo, Gil, and Corcho 2017; Stodden et al. 2016; Stodden and Miguez 2014; Gruening et al. 2018)</span>. This practice facilitates the preservation of methods utilized in data-intensive scientific workflows and enables verification of the published claims without requiring the end-user to do any manual installation and configuration. Examples of tools available via Docker containers used here are the alignment tool (BWA mem) used in the Alignment workflow and STAR aligner used in RNA-seq workflow.</p>
<h3 id="evaluating-provenance-profile">Evaluating Provenance Profile</h3>
<p>The retrospective provenance profile generated as part of <em>CWLProv</em> for each workflow enactment can be examined and queried to extract the required subset of information. <em>Provenance Analytics</em> is a separate domain and a next step after provenance collection in the provenance life cycle <span class="citation">(Missier 2016)</span>. Often provenance data is queried using specialized query languages such as SQL SPARQL or TriQL depending on the storage mechanism used. Query operations can combine information from prospective and retrospective provenance to understand computational experiments better.</p>
<p>The focus of this paper is not in-depth provenance analytics but we have demonstrated the application of the provenance profile generated as part of <em>CWLProv</em>. We have developed a commandline tool and Python API <em>“cwlprov-py”</em> <span class="citation">(Soiland-Reyes and Khan 2018)</span> for <em>CWLProv</em> RO analytics to interpret the captured retrospective provenance of CWL workflow enactment. This API currently supports the following use-cases.</p>
<p>Given a <em>CWLProv</em> RO:</p>
<ul>
<li><p><strong>Workflow Runs</strong> As each RO can contain more than one <em>workflow run</em> if sub-workflows are utilized to group related tasks into one workflow. In that case, the provenance traces are stored in separate files for each workflow run. <em>cwlprov-py</em> identifies the workflow enactments including the sub-workflows <em>(if any)</em> and returns the workflow identifiers annotated with the step names. The user can select the required trace and explore particular traces in detail.</p></li>
<li><p><strong>Attribution</strong> Each RO is assumed to be associated with a single enactment of the primary workflow and hence assumed to be enacted by one person. As discussed previously, <em>CWLProv</em> provides additional flags to enable user provenance capture. A user can provide their name and ORCID details that can be stored as part of a RO. <em>cwlprov-py</em> displays attribution details of the researcher responsible for the enactment <em>(if enabled)</em> and the versions of the workflow executor utilized in the analysis.</p></li>
<li><p><strong>Input/Output of a Process</strong> Provenance traces contain associations between the steps/workflows with the data they used or generated. A user interested in a particular step can identify the inputs used and outputs produced linked explicitly to that process using <em>cwlprov-py</em>. This option works using individual step identifiers (level 1) as well as nested workflows (level 2), facilitating re-use of intermediate data even if the original workflow author did not explicitly expose these as workflow outputs.</p></li>
<li><p><strong>Partial Re-runs</strong> Re-running or re-using only desired parts of a given workflow has been emphasized <span class="citation">(Cohen-Boulakia et al. 2017)</span> as important to evaluate the workflow process or validate the published results associated without necessarily re-enacting the workflow as a whole. <em>cwlprov-py</em> uses the identifier of the step/workflow to be re-run, parses the provenance trace to identify the inputs required and ultimately creates a JSON input object with the associated input parameters. This input object can then be used for partial re-runs of the desired step/workflow, making segmented analysis possible even for CWLProv consumers who don’t have sufficient hardware resources for re-executing more computationally heavy steps.</p></li>
</ul>
<p>While the above explores some use cases for consuming and re-using workflow execution data, we have not explored this in full detail. Further work could develop more specific user scenarios and perform usability testing with independent domain-experts who have not seen the executed workflow before.</p>
<p>An important point of <em>CWLProv</em> is to capture sufficient information at workflow execution time, so that post-processing (potentially by a third-party) can support unforeseen queries without requiring instrumentation at workflow design time. For instance, <code>cwlprov runtimes</code> calculates average runtime per step (requiring capture of start/stop time of each step iteration), while <code>cwlprov derived</code> calculates derivation paths back to input data (requiring consistent identifiers during execution). Further work could build a more researcher-oriented interface based on this approach, e.g. hardcoded data exploration for a particular workflow.</p>
<h3 id="temporal-and-spatial-overhead-with-provenance">Temporal and Spatial Overhead with Provenance</h3>
<p><span>|L | c | c | c | c | c | c |</span> <strong>Workflow</strong> &amp; &amp;<br />
&amp; &amp; <strong>toil-cwl-runner</strong> &amp; &amp; <strong>toil-cwl-runner</strong><br />
&amp; With Prov &amp; W/O Prov &amp; W/O Prov &amp; With Prov &amp; W/O Prov &amp; W/O Prov<br />
RNA-Seq Analysis Workflow &amp; 4m30.289s &amp; 4m0.139s &amp; 3m46.817s &amp; 3m33.306s &amp; 3m41.166s &amp; 3m30.406s<br />
Alignment Workflow&amp; 28m23.792s &amp; 24m12.404s &amp; 15m3.539s &amp; – &amp; 162m35.111s &amp; 146m27.592s<br />
Somatic Variant Calling Workflow &amp; 21m25.868s &amp; 19m27.519s &amp; 7m10.470s &amp; 17m26.722s &amp; 17m0.227s &amp; **<br />
<br />
</p>
<p>Table [tab:time] shows the run-times for the three workflow enactments using cwltool and toil-cwl-runner on Linux and MacOS with and without enabling provenance capture as described in the evaluation activity section. These workflows were enacted at least once before this time calculation, hence the timing does not include the time for Docker images to be downloaded. On a new system, when re-running these workflows for the first time, the Docker images will be downloaded and may take significantly longer than the time specified here especially in case of the Somatic Variant Calling workflow because of the image size.</p>
<p>Run-time and storage overheads are important for provenance-enabled computational experiments. The choice of different operating systems and provenance capture mechanisms such as operating-system level, application-level or workflow-level as well as I/O workload, interception mechanism and fine-grained information capture are key for provenance <span class="citation">(Carata et al. 2014; Kim and Vouk 2016)</span>.</p>
<p>In our case study, significant time difference can be seen for the alignment workflow that used the most voluminous dataset, hence producing a sizable RO as well. This was due to the RO-generation where data was aggregated within the RO. The difference between the provenance-enabled enactment versus the enactment without provenance is barely noticeable for the other two workflow enactments with the smaller datasets. The discussion about handling the big ‘-omics’ data such as human genome reference sequence, its index files and other database files (e.g. dbsnp) in Section <strong></strong> provides a possible solution to avoid such overheads.</p>
<p>In addition, noticeable time difference between the cwltool and toil-cwl-runner enactments is because of the default parallel versus serial job execution in case of <em>toil-cwl-runner</em> and <em>cwltool</em> respectively. The “scatter” operation in CWL when applied to one or more input parameters of a workflow step or a sub-workflow, supports parallel execution of the associated processes. Parallelism is also available without “scatter” when separate processes have all their inputs ready. If sufficient compute resources are available, these jobs will be enacted concurrently otherwise they are queued for subsequent execution. Compute intensive steps of a workflow can benefit from scatter features for parallel execution by reducing the overall run-time. Both Alignment and Somatic Variant Calling workflows utilize the scatter feature to enable higher degrees of parallel job execution in case of <em>toil-cwl-runner</em> which explains the time difference for the cross-executor of these two workflows. The difference is negligible for RNA-Seq workflow which is comprised of serial jobs with comparatively small test data.</p>
<h3 id="output-comparison-across-enactments">Output Comparison Across Enactments</h3>
<p>We compared the workflow outputs after each enactment to observe the concordance and/or discordance (if any) for the workflow enactment results produced across the platforms and across the executors. As <em>CWLProv</em> RO refers to the data with hashed checksums, these checksums are utilized for the result comparison. It is worth-mentioning that the comparison was made between the output files generated by the different enactments against a single <em>“truth-set”</em> output file available and checksum in the respective Git repositories.</p>
<p>The checksum of the output data generated cross-platform and cross-executor comparison data as a result of the initial enactments and re-runs using the CWL ROs to elicit the concordance in all but one cases. The “correctness” as well as agreement of these outputs given different execution environments (e.g. platform and executor) hold true except for Alignment workflow. Alignment workflow produced varying outputs after every execution even with the same executor and platform. The output of the alignment algorithm, “BWA mem” used in this workflow was non-deterministic as it depended on the <em>number of threads --t</em> and the <em>seed length --K</em> which affected the output produced. While the seed length in this case was set to a constant value, the number of threads varied depending on the availability of hardware resources at run-time, thereby resulting in varying output for the same input files.</p>
<h1 id="sec:discussion">Discussion and Future Directions</h1>
<p>This section discusses the current and future work with reference to enriched provenance capture and smart resource aggregation, and enhancements to both the <em>CWLProv</em> standard and implementation.</p>
<h3 id="compute-and-storage-resources">Compute and Storage Resources</h3>
<p>The <em>CWLProv</em> format encapsulates the data and workflow definitions involved in a given workflow enactment along with its retrospective provenance trace. CWL as a standard provides constructs to declare basic hardware resource requirements such as minimum and maximum cores, RAM and reserved file system storage required for a particular workflow enactment. The workflow authors can provide this information in the <em>“requirements”</em> or <em>“hints”</em> section as <em>“ResourceRequirement”</em>. These requirements/hints can be declared at workflow or individual step level, to help platforms/executors to allocate the required resources. This information indirectly stores some aspects of prospective view of provenance with respect to hardware requirements of the underlying system used to enact a workflow. Currently this information is only available if declared as part of workflow specification. In future, we plan to include these requirements as part of provenance for a given workflow such that all such information is gathered in one space and users are not required to inspect multiple sources to extract this information. This information can then be used as a pre-condition for potential successful enactment of a given workflow.</p>
<p>As <em>CWLProv</em> is focused on retrospective provenance capture of workflow enactment, we plan to include provenance information about the compute and storage resources utilized in a given enactment to fulfill <em>R18-resource-use</em>. We believe that documenting these resources will allow users to analyse their environment and resource allocations before execution, as opposed to trial and error methods that may result in multiple failed enactments of a given workflow. Despite being an important factor, it is surprising to see that most of existing provenance standards lack dedicated constructs to represent the underlying hardware resource usage information as part of prospective or retrospective provenance. In the case of complex workflows using distributed resources, where each step could be executed on a different node/server, including all this information in a single <em>PROV</em> profile will clutter the profile and render it potentially incomprehensible. Therefore, we plan to add a separate <em>Usage Record</em> document in the RO conforming to GFD.204 <span class="citation">(Cristofori et al. 2013)</span> to describe <em>Level 1</em> (and potentially <em>Level 2</em>) resource usage in a common format independent on actual execution environment.</p>
<p>Capturing such resource usage records require a tighter integration with the execution platform, and so we consider this future work better suited for a cloud-based CWL engine like <em>Toil</em> or <em>Arvados</em>, as the reference implementation <em>cwltool</em> does not exercise fine-grained control of its task execution. Detailed raw log files can also be provided as <em>Level 0</em> provenance, as we have demonstrated with cwltool, but these will by their nature be custom per execution platform and thus should be considered unstructured. Related work that is already exploring this approach is <em>cwl-metrics</em> <span class="citation">(Ohta, Tanjo, and Ogasawara 2018)</span>, which analyses raw <em>cwltool</em> log files in combination with detailed Docker invocation statistics using the container monitoring tool <em>Telegraf</em>. Ongoing collaboration is exploring adding these metrics as additional provenance to the <em>CWLProv</em> RO with summaries in PROV and GFD.204 formats.</p>
<h3 id="sec:domainknowledge">Provenance Profile Augmented with Domain Knowledge</h3>
<p><em>CWLProv</em> benefits from existing best practices proposed by numerous studies (Table [tab:recommendation:wide]) and includes defined standards for workflow representation, resource aggregation and provenance tracking (Section <strong></strong>). We posit that the principle of following well-defined data and metadata standards enables explicit data sharing and reuse. In order to include rich metadata for bioinformaticians to produce specialized ROs for bioinformatics to achieve <em>CWLProv</em> <em>Level 3</em> as defined in section <strong></strong>, we are investigating re-use of concepts from the BioCompute Object (BCO) project <span class="citation">(Alterovitz 2019)</span>. This domain-specific information is not necessary for computation and execution but for understandability of the shared resources. We encourage workflow authors to include such metadata and external identifiers for data and underlying tools, e.g. EDAM identifiers for the resources employed in designing a given workflow. The plan is to extract these annotations and represent in the retrospective provenance profile in <em>CWLProv</em> to ultimately achieve pragmatic interoperability by providing domain-specific scientific context of the experiments. Domain-specific information is essential in determining the nature of inputs, outputs and context of the processes linked to a given workflow enactment <span class="citation">(Alper et al. 2018)</span>. This information can be captured in the RO if and only if the workflow author adds it in the workflow definition, thus achieving <em>CWLProv</em> <em>Level 3</em> depends on the individual workflows.</p>
<h3 id="big--omics-data">Big -omics Data</h3>
<p>While aggregating all resources as one download-able object improves reproducibility, the size of the resulting RO is an important factor in practice. On one hand, completeness of the resources contributes towards minimizing the <em>workflow decay</em> phenomenon by least dependence on availability of third party resources. On the other hand, the nature of -omics data sizes can result in hard-to-manage workflow-centric ROs also leading to the spatial and temporal overheads as discussed in evaluation.</p>
<p>One solution is archiving the big datasets in online repositories or data stores and including the existing persistent identifiers and checksums in the RO instead of the actual data files, as previously demonstrated with BDBags <span class="citation">(Chard et al. 2016; Madduri 2019)</span>. While CWL executors like <em>toil-cwl-runner</em> can be configured to deposit data in a shared repository, the <em>cwltool</em> reference implementation explored in this study can only write to the local file system. External references raise the risk of unavailability of data at a later time. Therefore we recommend including the data in the RO if sufficient network and storage resources are available. Future work may explore post-processing <em>CWLProv</em> ROs to replace large data files with references to stable data repositories, producing a slimmer RO for transfer where individual data items can be retrieved on demand, as well as reducing data duplication across multiple related ROs.</p>
<h3 id="improving-cwlprov-efficiency-with-selective-provenance-capture">Improving <em>CWLProv</em> efficiency with selective provenance capture</h3>
<p><em>Shim</em> refers to an adaptor step to resolve a format incompatibility issues between two workflow tasks <span class="citation">(Mohan, Lu, and Kotov 2014)</span>, typically converting the previous output into an acceptable format for the next step. For example in our case study <em>RNA-seq</em> workflow, <em>RNA-SeQC</em> require an indexed BAM file, whereas the output of <em>STAR</em> or <em>Picard MarkDuplicates</em> only comprises of the BAM file alone. Hence, a shim step executing <em>SAMtools index</em> make the aligned reads analysis ready for RNA-SeQC. Compared to the more analytical steps, the provenance of such shim steps are not particularly interesting for domain scientists, and in many cases their intermediate data would effectively double the storage cost with little information gains, as such data can be reliably recreated by re-applying the predictable transformation step (considering it as a <em>pure function</em> without side-effects). Another type of ignorable steps could be purely diagnostic, which outputs are used primarily during workflow design to verify tool settings. A workflow engine does not necessarily know which steps are “boring”<a href="#fn3" class="footnoteRef" id="fnref3"><sup>3</sup></a> and our proof of concept implementation will dutifully store provenance from all steps.</p>
<p>To improve efficiency, future <em>CWLProv</em> work could add options to ignore capturing outputs of specified <em>shim</em> steps, or to not store files over a particular file size. Similarly a scientist or a WMS may elect to only capture provenance at a particular provenance level (see Section <strong></strong>). Provenance captured under such settings would be “incomplete” (e.g. PROV would say <em>RNA-SeQC</em> consumed an identified BAM index file, but the corresponding bytes would not be stored in the RO), thus it is envisioned this can be indicated in the RO manifest as a variant of the <em>CWLProv</em> profile identifier to give the end-user clear indication of what to expect in terms of completeness, so that tools like <em>cwlprov-py</em> could be extended to re-create missing outputs, verifying their expected checksums, or collapse provenance listing of “boring” steps to improve human tractability.</p>
<h3 id="enforcement-of-best-practices-an-open-problem"> Enforcement of Best Practices – An Open Problem</h3>
<p>Recommendations and best practices from the scientific community are proposed frequently, to guide researchers to design their computational experiments in such a way as to make their research reproducible and verifiable. Not only the best practices for workflow design, but also for resource declaration, software packaging and configuration management are put forward <span class="citation">(Gruening et al. 2018)</span> to avoid dependence on local installations and manual processes of dependency management. The term <em>“Better Software, Better Research”</em> <span class="citation">(Goble 2014)</span> can also be well-applied on and adapted for the workflow design process.</p>
<p>Declarative approaches to workflow definition such as CWL facilitate and encourage users to explicitly declare everything in a workflow, improving white-box view of the retrospective as well as prospective provenance. Such workflows should provide insights of the complete process followed, to produce a data artefact resolving the black-boxness often associated with the workflow provenance. However, it is entirely up to researchers to leverage these approaches to produce well-defined workflows with explicit details facilitating enriched capture of the provenance trace at the appropriate level, and this can require considerable effort and consistency on the workflow designer’s behalf. For instance, the alignment workflow used in this case study embeds bash scripts into the CWL tool definition, therefore requiring another layer needed to be penetrated for provenance information extraction. Despite using CWL for the workflow definition and <em>CWLProv</em> for provenance capture, the provenance trace will be missing critical information making it coarse-grained, and the raw logs capturing the enactment will also not be as informative.</p>
<p>The three criteria defined by <span class="citation">Cohen-Boulakia et al. (2017)</span> to be followed by workflow designers are: modularized specifications, unified representation and workflow annotations. CWL facilitates a modular structure to workflow definitions by coupling similar steps to <em>subworkflows</em>; and, as an interoperable standard, CWL provides a common platform moving towards resolution of the heterogeneity of the workflow specification languages. In addition, users can add standardised domain-specific annotations to data and workflows incorporating the constructs defined by external ontologies (e.g. EDAM) to enhance understanding of the shared specification and the resources it refers to. All these features can be utilized to design better workflows and maximize the information declaration resulting in semantically-rich and provenance-complete <em>CWLProv</em> ROs, and should thus be expressed clearly in user guides<a href="#fn4" class="footnoteRef" id="fnref4"><sup>4</sup></a> for workflow authors.</p>
<p>The usability of any <em>CWLProv</em> RO directly relies on the choice of practices followed by the researchers to design and communicate their computational analyses. Workflow-centric initiatives similar to <em>software carpentry</em> <span class="citation">(“Software Carpentry” 4AD–10AD)</span> and <em>code is science</em> <span class="citation">(“Code Is Science” 4AD–10AD)</span> are one possible way to organize training and create awareness around best practices. Community-driven efforts to further consolidate the understanding of requirements to make a given workflow explicit and understandable should be made. Not only awareness about the workflow design is needed, but also the availability of the associated resources should be emphasized e.g. software as containers or software packages, big datasets in public repositories and pre-processing/post-processing as part of workflow. Without putting proposed best practices into actual practice, complete communication and hence the reproducibility of a workflow-centric computational analysis is likely to remain challenging.</p>
<h1 id="sec:conclusion">Conclusion</h1>
<p>The comprehensive sharing and communication of the computational experiments employed to achieve a scientific objective establishes trust on published results. Shared resources are sometimes rendered ineffective due to incomplete provenance, heterogeneity of platforms, unavailability of software and limited access to data. To this context, the contributions of this study are four-fold. First, we have provided a comprehensive summary of the recommendations put forward by the community regarding workflow design and resource sharing. Second, we define a hierarchical provenance framework to achieve homogeneity in the granularity of the information shared with each level addressing specific provenance recommendations.</p>
<p>Third, we leverage the existing standards best suited to define a standardized format, <em>CWLProv</em> for methodical representation of workflow enactments, its provenance and the associated artefacts employed. Finally, to demonstrate the applicability of <em>CWLProv</em>, we extend an existing workflow executor (<em>cwltool</em>) to provide a reference implementation to generate interoperable workflow-centric ROs, aggregating and preserving data and methods to support the coherent sharing of computational analyses and experiments.</p>
<p>With any published scientific research, statements such as <em>“Methods and data are available upon request”</em> should no longer be acceptable in a modern open-science-driven research community. Considering on one hand the collaborative nature and emerging openness of bioinformatics research and on the other hand the heterogeneity of workflow design approaches, it is essential to provide open access to the structured representation of the data and methods utilized in any scientific study to achieve interoperable solutions facilitating reproducibility of science.</p>
<p>Provenance capture and its subsequent use to support published research transparency should not be treated as an after-thought but rather as a standard practice of up-most priority. With adoption of well-defined standards for provenance and declarative workflow definition approaches, the assumption of black-box provenance often associated with workflows can be addressed. The workflow authors should be encouraged to follow well-established and agreed upon best practices for workflow design and software environment deployment. In conclusion, we do not require new standards, new WMSs or indeed new best practices, instead the focus should be to implement, utilize and re-use existing mature community-driven initiatives to achieve consensus in representing different aspects of computational experiments.</p>
<h1 id="availability-of-source-code-and-requirements">Availability of source code and requirements</h1>
<p><em>CWLProv</em> is implemented as part of the CWL reference implementation <em>cwltool</em>:</p>
<ul>
<li><p>Project name: cwltool <span> (<a href="http://identifiers.org/rrid/RRID:RRID:SCR_015528"></a>) </span></p></li>
<li><p>Project home page: <a href="https://github.com/common-workflow-language/cwltool" class="uri">https://github.com/common-workflow-language/cwltool</a></p></li>
<li><p>Version: <a href="https://pypi.org/project/cwltool/1.0.20180809224403/">1.0.20181012180214</a> <span class="citation">(Amstutz et al. 2018)</span></p></li>
<li><p>Operating system(s): Platform independent</p></li>
<li><p>Programming language: Python 3.5 or later <span> (<a href="http://identifiers.org/rrid/RRID:RRID:SCR_008394"></a>) </span></p></li>
<li><p>Other requirements: <a href="https://www.docker.com/">Docker</a> <span> (<a href="http://identifiers.org/rrid/RRID:RRID:SCR_016445"></a>) </span> recommended</p></li>
<li><p>License: <a href="http://www.apache.org/licenses/LICENSE-2.0">Apache License, Version 2.0</a></p></li>
</ul>
<p>The <em><em>CWLProv</em> profile</em> documents the use of W3C PROV in a Research Object to capture a CWL workflow run:</p>
<ul>
<li><p>Project name: <em>CWLProv</em> profile</p></li>
<li><p>Project home page: <a href="https://w3id.org/cwl/prov" class="uri">https://w3id.org/cwl/prov</a></p></li>
<li><p>Version: <a href="https://w3id.org/cwl/prov/0.6.0">0.6.0</a> <span class="citation">(Soiland-Reyes, Khan, and Crusoe 2018)</span></p></li>
<li><p>Operating system(s): Platform independent</p></li>
<li><p>License: <a href="http://www.apache.org/licenses/LICENSE-2.0">Apache License, Version 2.0</a></p></li>
</ul>
<p>The <em><em>CWLProv</em> Python Tool</em> can be used to explore <em>CWLProv</em> ROs on the command line:</p>
<ul>
<li><p>Project name: <em>CWLProv</em> Python Tool (<em>cwlprov-py</em>)</p></li>
<li><p>Project home page: <a href="https://github.com/common-workflow-language/cwlprov-py" class="uri">https://github.com/common-workflow-language/cwlprov-py</a></p></li>
<li><p>Version: <a href="https://pypi.org/project/cwlprov/0.1.1/">0.1.1</a> <span class="citation">(Soiland-Reyes and Khan 2018)</span></p></li>
<li><p>Operating system(s): Platform independent</p></li>
<li><p>Programming language: Python 3.5 or later <span> (<a href="http://identifiers.org/rrid/RRID:RRID:SCR_008394"></a>) </span></p></li>
<li><p>License: <a href="http://www.apache.org/licenses/LICENSE-2.0">Apache License, Version 2.0</a></p></li>
</ul>
<h1 id="availability-of-supporting-data-and-materials">Availability of supporting data and materials</h1>
<p><em>CWLProv</em> Research Objects of CWL workflow executions are published in Mendeley Data and mirrored to Zenodo.</p>
<ul>
<li><p>CWL run of Somatic Variant Calling Workflow (CWLProv 0.5.0 Research Object) <span class="citation">(Khan and Soiland-Reyes 2018c)</span><br />
<a href="https://doi.org/10.17632/97hj93mkfd.3" class="uri">https://doi.org/10.17632/97hj93mkfd.3</a><br />
<a href="https://zenodo.org/record/2841641" class="uri">https://zenodo.org/record/2841641</a></p></li>
<li><p>CWL run of Alignment Workflow (CWLProv 0.6.0 Research Object) <span class="citation">(Khan and Soiland-Reyes 2018a)</span><br />
<a href="https://doi.org/10.17632/6wtpgr3kbj.1" class="uri">https://doi.org/10.17632/6wtpgr3kbj.1</a><br />
<a href="https://zenodo.org/record/2632836" class="uri">https://zenodo.org/record/2632836</a></p></li>
<li><p>CWL run of RNA-seq Analysis Workflow (CWLProv 0.5.0 Research Object) <span class="citation">(Khan and Soiland-Reyes 2018b)</span><br />
<a href="https://doi.org/10.17632/xnwncxpw42.1" class="uri">https://doi.org/10.17632/xnwncxpw42.1</a><br />
<a href="https://zenodo.org/record/2838898" class="uri">https://zenodo.org/record/2838898</a></p></li>
</ul>
<p>The <em>CWLProv Python Tool</em> can be used to explore the above research objects.</p>
<h1 id="declarations">Declarations</h1>
<h2 id="list-of-abbreviations">List of abbreviations</h2>
<p>BAM: Binary Alignment Map; BCO: BioCompute Object; CRAM: Compressed Alignment Map; CWL: Common Workflow Language; EBI: European Bionformatics Institute; GATK: Genome Analysis ToolKit; HPC: High Performance Computing; JSON-LD: JavaScript Object Notation for Linked Data; OS: Operating System; PROV-DM: PROVenance Data Model; RO: Research Object; W3C: World Wide Web Consortium; WMS: Workflow Management System;</p>
<h2 id="ethical-approval-optional">Ethical Approval (optional)</h2>
<p>Not applicable.</p>
<h2 id="consent-for-publication">Consent for publication</h2>
<p>Not applicable.</p>
<h2 id="competing-interests">Competing Interests</h2>
<p>SSR and MRC are members of the leadership team for Common Workflow Language at the Software Freedom Conservancy.</p>
<h2 id="funding">Funding</h2>
<p>FZK funded by Melbourne International Research Scholarship (MIRS) and Melbourne International Fee Remission Scholarship (MIFRS).</p>
<p>SSR and CG funded by <a href="https://www.bioexcel.eu">BioExcel CoE</a>, a project funded by the European Commission <a href="http://dx.doi.org/10.13039/100010666">Horizon 2020 Framework Programme</a> under contracts <a href="https://cordis.europa.eu/project/id/823830">H2020-INFRAEDI-02-2018-823830</a> and <a href="http://cordis.europa.eu/projects/675728">H2020-EINFRA-2015-1-675728</a>, as well as <a href="https://www.ibisba.eu/">IBISBA</a> (<a href="http://cordis.europa.eu/projects/730976">H2020-INFRAIA-1-2014-2015-730976</a>).</p>
<h2 id="authors-contributions">Author’s Contributions</h2>
<p>Conceptualization: FZK, SSR, MRC. Data curation: FZK. Formal analysis: FZK. Funding acquisition: ROS, AL, CAG. Investigation: FZK. Methodology: FZK, SSR. Project administration: FZK, SSR, ROS, AL Computing Resources: ROS, AL. Software: FZK, SSR, MRC. Supervision: MRC, ROS, AL, CAG. Validation: FZK, SSR. Writing - original draft: FZK. Writing - review &amp; editing: FZK, SSR, ROS, AL, MRC.</p>
<h1 id="acknowledgements">Acknowledgements</h1>
<p>An earlier version of this article <span class="citation">(Khan et al. 2018)</span> was submitted for consideration at International Provenance and Annotation Workshop (IPAW) 2018. We would like to thank the IPAW reviewers for their constructive comments.</p>
<p>We would also like to thank the GigaScience editors and reviewers Tomoya Tanjo and Alban Gaignard for constructive and valuable feedback we think has improved the manuscript and future directions.</p>
<p>We would like to thank the Common Workflow Language community, and in particular Peter Amstutz, Pau Ruiz Safont and Pjotr Prins, for their continuing support, review and feedback. We would also like to thank Brad Chapman, Christopher Ball and Lon Blauvelt for the workflows used in the evaluation and their prompt replies to our enquiries.</p>
<p>We are grateful for partial travel support from Open Bioinformatics Foundation (OBF) Travel Fellowship Program <span class="citation">(“OBF Travel Fellowship Program | Obf News” 2016)</span> to Farah Zaib Khan for attending the Bioinformatics Open Source Conference (BOSC) 2017 and 2018 Codefests subsidizing this collaborative effort.</p>
<div id="refs" class="references">
<div id="ref-Abecasis14">
<p>“Abecasis Lab - Genome Analysis Wiki.” 2017. <a href="https://genome.sph.umich.edu/wiki/Abecasis_Lab" class="uri">https://genome.sph.umich.edu/wiki/Abecasis_Lab</a>.</p>
</div>
<div id="ref-Alper2018">
<p>Alper, Pinar, Khalid Belhajjame, Vasa Curcin, and Carole Goble. 2018. “LabelFlow Framework for Annotating Workflow Provenance.” <em>Informatics</em> 5 (1). MDPI AG: 11. doi:<a href="https://doi.org/10.3390/informatics5010011">10.3390/informatics5010011</a>.</p>
</div>
<div id="ref-Alterovitz2019">
<p>Alterovitz, Dennis AND Goble, Gil AND Dean. 2019. “Enabling Precision Medicine via Standard Communication of Hts Provenance, Analysis, and Results.” <em>PLOS Biology</em> 16 (12). Public Library of Science: 1–14. doi:<a href="https://doi.org/10.1371/journal.pbio.3000099">10.1371/journal.pbio.3000099</a>.</p>
</div>
<div id="ref-AmazonEC">
<p>“Amazon Ec2.” 1AD–12AD. <a href="https://aws.amazon.com/ec2/" class="uri">https://aws.amazon.com/ec2/</a>.</p>
</div>
<div id="ref-cwltool">
<p>Amstutz, Peter, Michael R. Crusoe, Farah Zaib Khan, Stian Soiland-Reyes, Manvendra Singh, Kapil kumar, Anton Khodak, et al. 2018. “common-workflow-language/cwltool: 1.0.20181012180214.” Zenodo. doi:<a href="https://doi.org/10.5281/zenodo.1471589">10.5281/zenodo.1471589</a>.</p>
</div>
<div id="ref-cwl">
<p>Amstutz, Peter, Michael R. Crusoe, Nebojša Tijanić, Brad Chapman, John Chilton, Michael Heuer, Andrey Kartashov, et al. 2016. “Common Workflow Language, V1.0.” Figshare. doi:<a href="https://doi.org/10.6084/m9.figshare.3115156.v2">10.6084/m9.figshare.3115156.v2</a>.</p>
</div>
<div id="ref-Angiuoli_2011">
<p>Angiuoli, Samuel V., James R. White, Malcolm Matalka, Owen White, and W. Florian Fricke. 2011. “Resources and Costs for Microbial Sequence Analysis Evaluated Using Virtual Machines and Cloud Computing.” Edited by Sarah K.Editor Highlander. <em>PLoS ONE</em> 6 (10). Public Library of Science (PLoS): e26624. doi:<a href="https://doi.org/10.1371/journal.pone.0026624">10.1371/journal.pone.0026624</a>.</p>
</div>
<div id="ref-arvados">
<p>“Arvados - Open Source Big Data Processing and Bioinformatics.” 8AD–11AD. <a href="https://arvados.org/" class="uri">https://arvados.org/</a>.</p>
</div>
<div id="ref-atkinson_2017">
<p>Atkinson, Malcolm, Sandra Gesing, Johan Montagnat, and Ian Taylor. 2017. “Scientific Workflows: Past, Present and Future.” <em>Future Generation Computer Systems</em> 75 (October): 216–27. doi:<a href="https://doi.org/10.1016/j.future.2017.05.041">10.1016/j.future.2017.05.041</a>.</p>
</div>
<div id="ref-bechhofer_2013">
<p>Bechhofer, Sean, Iain Buchan, David De Roure, Paolo Missier, John Ainsworth, Jiten Bhagat, Philip Couch, et al. 2013. “Why Linked Data Is Not Enough for Scientists.” <em>Future Gener. Comput. Syst.</em> 29 (2): 599–611.</p>
</div>
<div id="ref-belhajjame_2012">
<p>Belhajjame, Khalid, Oscar Corcho, Daniel Garijo, Jun Zhao, Paolo Missier, David Newman, Raúl Palma, et al. 2012. “Workflow-Centric Research Objects: First Class Citizens in Scholarly Discourse.” In <em>Proceedings of the 2nd Workshop on Semantic Publishing (SePublica 2012)</em>, 903:1–12. CEUR Workshop Proceedings. <a href="http://ceur-ws.org/Vol-903/paper-01.pdf" class="uri">http://ceur-ws.org/Vol-903/paper-01.pdf</a>.</p>
</div>
<div id="ref-belhajjame_2015">
<p>Belhajjame, Khalid, Jun Zhao, Daniel Garijo, Matthew Gamble, Kristina Hettne, Raul Palma, Eleni Mina, et al. 2015. “Using a Suite of Ontologies for Preserving Workflow-Centric Research Objects.” <em>Web Semantics: Science, Services and Agents on the World Wide Web</em> 32 (0): 16–42. doi:<a href="https://doi.org/10.1016/j.websem.2015.01.003">10.1016/j.websem.2015.01.003</a>.</p>
</div>
<div id="ref-benabdelkader_2015">
<p>Benabdelkader, Ammar, Antoine Ahc VanKampen, and Silvia D Olabarriaga. 2015. “PROV-Man: A PROV-Compliant Toolkit for Provenance Management.” <em>PeerJ PrePrints</em> 3 (e1102v1). PeerJ, Inc. doi:<a href="https://doi.org/10.7287/peerj.preprints.1102v1">10.7287/peerj.preprints.1102v1</a>.</p>
</div>
<div id="ref-biosimspace">
<p>“BioSimSpace.” 2018. <a href="https://biosimspace.org/" class="uri">https://biosimspace.org/</a>.</p>
</div>
<div id="ref-bcbio">
<p>“Blue Collar Bioinformatics.” <a href="http://bcb.io/" class="uri">http://bcb.io/</a>.</p>
</div>
<div id="ref-Bubak_2013">
<p>Bubak, M., M. Kasztelnik, M. Malawski, J. Meizner, P. Nowakowski, and S. Varma. 2013. “Evaluation of Cloud Providers for Vph Applications.” <em>2013 13th IEEE/ACM International Symposium on Cluster, Cloud, and Grid Computing</em>, May. IEEE. doi:<a href="https://doi.org/10.1109/ccgrid.2013.54">10.1109/ccgrid.2013.54</a>.</p>
</div>
<div id="ref-caoprovone">
<p>Cao, Yang, Christopher Jones, Victor Cuevas-Vicenttin, Matthew B Jones, Bertram Ludäscher, Timothy McPhillips, Paolo Missier, et al. 2016. “ProvONE: Extending Prov to Support the Dataone Scientific Community,” June. <a href="http://homepages.cs.ncl.ac.uk/paolo.missier/doc/dataone-prov-3-years-later.pdf" class="uri">http://homepages.cs.ncl.ac.uk/paolo.missier/doc/dataone-prov-3-years-later.pdf</a>.</p>
</div>
<div id="ref-Carata2014">
<p>Carata, Lucian, Sherif Akoush, Nikilesh Balakrishnan, Thomas Bytheway, Ripduman Sohan, Margo Selter, and Andy Hopper. 2014. “A Primer on Provenance.” <em>Communications of the ACM</em> 57 (5). Association for Computing Machinery (ACM): 52–60. doi:<a href="https://doi.org/10.1145/2596628">10.1145/2596628</a>.</p>
</div>
<div id="ref-Casati1998">
<p>Casati, F, S Ceri, B Pernici, and G Pozzi. 1998. “Workflow Evolution.” <em>Data &amp; Knowledge Engineering</em> 24 (3). Elsevier BV: 211–38. doi:<a href="https://doi.org/10.1016/s0169-023x(97)00033-5">10.1016/s0169-023x(97)00033-5</a>.</p>
</div>
<div id="ref-nextflow_tracing">
<p>Centre for Genomic Regulation (CRG). 2018. “Nextflow: Tracing and Visualization.” <a href="https://www.nextflow.io/docs/latest/tracing.html#trace-report" class="uri">https://www.nextflow.io/docs/latest/tracing.html#trace-report</a>.</p>
</div>
<div id="ref-chard_2016">
<p>Chard, Kyle, Mike D’Arcy, Ben Heavner, Ian Foster, Carl Kesselman, Ravi Madduri, Alexis Rodriguez, et al. 2016. “I’ll Take That to Go: Big Data Bags and Minimal Identifiers for Exchange of Large, Complex Datasets.” In <em>2016 IEEE International Conference on Big Data (Big Data)</em>, 319–28. IEEE. doi:<a href="https://doi.org/10.1109/bigdata.2016.7840618">10.1109/bigdata.2016.7840618</a>.</p>
</div>
<div id="ref-chen2011partitioning">
<p>Chen, Weiwei, and Ewa Deelman. 2011. “Partitioning and Scheduling Workflows Across Multiple Sites with Storage Constraints.” In <em>International Conference on Parallel Processing and Applied Mathematics</em>, 11–20. Springer. doi:<a href="https://doi.org/10.1007/978-3-642-31500-8_2">10.1007/978-3-642-31500-8_2</a>.</p>
</div>
<div id="ref-Chirigati2016">
<p>Chirigati, Fernando, Rémi Rampin, Dennis Shasha, and Juliana Freire. 2016. “ReproZip.” In <em>Proceedings of the 2016 International Conference on Management of Data - SIGMOD 16</em>. ACM Press. doi:<a href="https://doi.org/10.1145/2882903.2899401">10.1145/2882903.2899401</a>.</p>
</div>
<div id="ref-Clark2014">
<p>Clark, Tim, Paolo N Ciccarese, and Carole A Goble. 2014. “Micropublications: A Semantic Model for Claims, Evidence, Arguments and Annotations in Biomedical Communications.” <em>Journal of Biomedical Semantics</em> 5 (1). Springer Nature: 28. doi:<a href="https://doi.org/10.1186/2041-1480-5-28">10.1186/2041-1480-5-28</a>.</p>
</div>
<div id="ref-clifford_2008">
<p>Clifford, Ben, Ian Foster, Jens-S Voeckler, Michael Wilde, and Yong Zhao. 2008. “Tracking Provenance in a Virtual Data Grid.” <em>Concurrency and Computation: Practice and Experience</em> 20 (5). John Wiley &amp; Sons, Ltd.: 565–75. doi:<a href="https://doi.org/10.1002/cpe.1256">10.1002/cpe.1256</a>.</p>
</div>
<div id="ref-Cochrane2012">
<p>Cochrane, Guy, Blaise Alako, Clara Amid, Lawrence Bower, Ana Cerdeño-Tárraga, Iain Cleland, Richard Gibson, et al. 2012. “Facing Growth in the European Nucleotide Archive.” <em>Nucleic Acids Research</em> 41 (D1). Oxford University Press (OUP): D30–D35. doi:<a href="https://doi.org/10.1093/nar/gks1175">10.1093/nar/gks1175</a>.</p>
</div>
<div id="ref-CodeasaR">
<p>“Code as a Research Object.” <a href="http://mozillascience.github.io/code-research-object/" class="uri">http://mozillascience.github.io/code-research-object/</a>.</p>
</div>
<div id="ref-CodeIsSc">
<p>“Code Is Science.” 4AD–10AD. <a href="http://www.codeisscience.com/" class="uri">http://www.codeisscience.com/</a>.</p>
</div>
<div id="ref-cohen2017scientific">
<p>Cohen-Boulakia, Sarah, Khalid Belhajjame, Olivier Collin, Jérôme Chopard, Christine Froidevaux, Alban Gaignard, Konrad Hinsen, et al. 2017. “Scientific Workflows for Computational Reproducibility in the Life Sciences: Status, Challenges and Opportunities.” <em>Future Generation Computer Systems</em> 75 (October): 284–98. doi:<a href="https://doi.org/10.1016/j.future.2017.01.012">10.1016/j.future.2017.01.012</a>.</p>
</div>
<div id="ref-cwl-existing-workflow-systems">
<p>Common Workflow Language project. 2018. “Existing Workflow Systems.” Edited by CWL Community. Common Workflow Language project. <a href="https://s.apache.org/existing-workflow-systems" class="uri">https://s.apache.org/existing-workflow-systems</a>.</p>
</div>
<div id="ref-cwltool-controlflow">
<p>“Common-Workflow-Language/Cwltool:Common Workflow Language Reference Implementation.” 2016. <a href="https://github.com/common-workflow-language/cwltool#cwl-tool-control-flow" class="uri">https://github.com/common-workflow-language/cwltool#cwl-tool-control-flow</a>.</p>
</div>
<div id="ref-Conda">
<p>“Conda Conda Documentation.” <a href="https://conda.io/docs/" class="uri">https://conda.io/docs/</a>.</p>
</div>
<div id="ref-Conesa2016">
<p>Conesa, Ana, Pedro Madrigal, Sonia Tarazona, David Gomez-Cabrero, Alejandra Cervera, Andrew McPherson, Michał Wojciech Szcześniak, et al. 2016. “A Survey of Best Practices for RNA-Seq Data Analysis.” <em>Genome Biology</em> 17 (1). Springer Nature. doi:<a href="https://doi.org/10.1186/s13059-016-0881-8">10.1186/s13059-016-0881-8</a>.</p>
</div>
<div id="ref-cristofori2013usage">
<p>Cristofori, A, IGI Bologna, J Gordon, STFC RAL London, JA Kennedy, RZG Munich, and R Müller-Pfefferkorn. 2013. “Usage Record–Format Recommendation.” In <em>Open Grid Forum</em>. <a href="https://www.ogf.org/documents/GFD.204.pdf" class="uri">https://www.ogf.org/documents/GFD.204.pdf</a>.</p>
</div>
<div id="ref-cuevasvicenttn_2012">
<p>Cuevas-Vicenttín, Víctor, Saumen Dey, Sven Köhler, Sean Riddle, and Bertram Ludäscher. 2012. “Scientific Workflows and Provenance: Introduction and Research Opportunities.” <em>Datenbank Spektrum</em> 12 (3): 193–203. doi:<a href="https://doi.org/10.1007/s13222-012-0100-z">10.1007/s13222-012-0100-z</a>.</p>
</div>
<div id="ref-Custovic799">
<p>Custovic, Adnan, John Ainsworth, Hasan Arshad, Christopher Bishop, Iain Buchan, Paul Cullinan, Graham Devereux, et al. 2015. “The Study Team for Early Life Asthma Research (Stelar) Consortium Asthma E-Lab: Team Science Bringing Data, Methods and Investigators Together.” <em>Thorax</em> 70 (8). BMJ Publishing Group Ltd: 799–801. doi:<a href="https://doi.org/10.1136/thoraxjnl-2015-206781">10.1136/thoraxjnl-2015-206781</a>.</p>
</div>
<div id="ref-DataBios21">
<p>“Data Biosphere.” 2018. <a href="https://github.com/DataBioSphere" class="uri">https://github.com/DataBioSphere</a>.</p>
</div>
<div id="ref-DataComm86">
<p>“Data Commons | Nih Common Fund.” 2018. <a href="https://commonfund.nih.gov/commons" class="uri">https://commonfund.nih.gov/commons</a>.</p>
</div>
<div id="ref-DeLuca2012">
<p>DeLuca, David S., Joshua Z. Levin, Andrey Sivachenko, Timothy Fennell, Marc-Danie Nazaire, Chris Williams, Michael Reich, Wendy Winckler, and Gad Getz. 2012. “RNA-SeQC: RNA-Seq Metrics for Quality Control and Process Optimization.” <em>Bioinformatics</em> 28 (11). Oxford University Press (OUP): 1530–2. doi:<a href="https://doi.org/10.1093/bioinformatics/bts196">10.1093/bioinformatics/bts196</a>.</p>
</div>
<div id="ref-ditommaso_2017">
<p>Di Tommaso, Paolo, Maria Chatzou, Evan W Floden, Pablo Prieto Barja, Emilio Palumbo, and Cedric Notredame. 2017. “Nextflow Enables Reproducible Computational Workflows.” <em>Nat Biotechnol</em> 35 (4): 316–19. doi:<a href="https://doi.org/10.1038/nbt.3820">10.1038/nbt.3820</a>.</p>
</div>
<div id="ref-DigitalOcean">
<p>“DigitalOcean - Cloud Computing, Simplicity at Scale.” 1AD–12AD. <a href="https://www.digitalocean.com/" class="uri">https://www.digitalocean.com/</a>.</p>
</div>
<div id="ref-dobin2015mapping">
<p>Dobin, Alexander, and Thomas R Gingeras. 2015. “Mapping Rna-Seq Reads with Star.” <em>Current Protocols in Bioinformatics</em> 51 (1). Wiley Online Library: 11–14. doi:<a href="https://doi.org/10.1002/0471250953.bi1114s51">10.1002/0471250953.bi1114s51</a>.</p>
</div>
<div id="ref-Dobin2012">
<p>Dobin, Alexander, Carrie A. Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R. Gingeras. 2012. “STAR: Ultrafast Universal RNA-Seq Aligner.” <em>Bioinformatics</em> 29 (1). Oxford University Press (OUP): 15–21. doi:<a href="https://doi.org/10.1093/bioinformatics/bts635">10.1093/bioinformatics/bts635</a>.</p>
</div>
<div id="ref-services_2012">
<p>EMC Education Services. 2012. <em>Information Storage and Management: Storing, Managing, and Protecting Digital Information in Classic, Virtualized, and Cloud Environments, Second Edition</em>. Edited by Somasundaram Gnanasundaram and Alok Shrivastava. John Wiley &amp; Sons.</p>
</div>
<div id="ref-samblaster">
<p>Faust, G. G., and I. M. Hall. 2014. “SAMBLASTER: Fast Duplicate Marking and Structural Variant Read Extraction.” <em>Bioinformatics</em> 30 (17). Oxford University Press (OUP): 2503–5. doi:<a href="https://doi.org/10.1093/bioinformatics/btu314">10.1093/bioinformatics/btu314</a>.</p>
</div>
<div id="ref-figshare">
<p>“Figshare - Credit for All Your Research.” <a href="https://figshare.com/" class="uri">https://figshare.com/</a>.</p>
</div>
<div id="ref-freire_2012">
<p>Freire, Juliana, and Claudio T. Silva. 2012. “Making Computations and Publications Reproducible with VisTrails.” <em>Comput Sci Eng</em> 14 (4): 18–25. doi:<a href="https://doi.org/10.1109/{MCSE}.2012.76">10.1109/{MCSE}.2012.76</a>.</p>
</div>
<div id="ref-Gaignard2014">
<p>Gaignard, Alban, Johan Montagnat, Bernard Gibaud, Germain Forestier, and Tristan Glatard. 2014. “Domain-Specific Summarization of Life-Science E-Experiments from Provenance Traces.” <em>Journal of Web Semantics</em> 29 (December). Elsevier BV: 19–30. doi:<a href="https://doi.org/10.1016/j.websem.2014.07.001">10.1016/j.websem.2014.07.001</a>.</p>
</div>
<div id="ref-PoeM">
<p>Gaignard, Alban, Hala Skaf-Molli, and Audrey Bihouée. 2016. “From Scientific Workflow Patterns to 5-Star Linked Open Data.” In <em>TaPP</em>.</p>
</div>
<div id="ref-garijo_2017">
<p>Garijo, D, Y Gil, and O Corcho. 2017. “Abstract, Link, Publish, Exploit: An End to End Framework for Workflow Sharing.” <em>Future Generation Computer Systems</em> 75 (October): 271–83. doi:<a href="https://doi.org/10.1016/j.future.2017.01.008">10.1016/j.future.2017.01.008</a>.</p>
</div>
<div id="ref-garijo_2014">
<p>Garijo, Daniel, Yolanda Gil, and Oscar Corcho. 2014. “Towards Workflow Ecosystems Through Semantic and Standard Representations.” In <em>2014 9th Workshop on Workflows in Support of Large-Scale Science</em>, 94–104. IEEE. doi:<a href="https://doi.org/10.1109/{WORKS}.2014.13">10.1109/{WORKS}.2014.13</a>.</p>
</div>
<div id="ref-garijo_2013">
<p>Garijo, Daniel, Sarah Kinnings, Li Xie, Lei Xie, Yinliang Zhang, Philip E Bourne, and Yolanda Gil. 2013. “Quantifying Reproducibility in Computational Biology: The Case of the Tuberculosis Drugome.” <em>PLoS ONE</em> 8 (11): e80278. doi:<a href="https://doi.org/10.1371/journal.pone.0080278">10.1371/journal.pone.0080278</a>.</p>
</div>
<div id="ref-GATKBP">
<p>“GATK | Bp Doc #11165 | Data Pre-Processing for Variant Discovery.” 2018. <a href="https://software.broadinstitute.org/gatk/best-practices/workflow?id=11165" class="uri">https://software.broadinstitute.org/gatk/best-practices/workflow?id=11165</a>.</p>
</div>
<div id="ref-giesler_2017">
<p>Giesler, André, Myriam Czekala, Björn Hagemeier, and Richard Grunzke. 2017. “Uniprov: A Flexible Provenance Tracking System for UNICORE.” In <em>High-Performance Scientific Computing</em>, edited by Edoardo Di Napoli, Marc-André Hermanns, Hristo Iliev, Andreas Lintermann, and Alexander Peyser, 10164:233–42. Lecture Notes in Computer Science. Cham: Springer International Publishing. doi:<a href="https://doi.org/10.1007/978-3-319-53862-4\_20">10.1007/978-3-319-53862-4\_20</a>.</p>
</div>
<div id="ref-wings2011">
<p>Gil, Y., V. Ratnakar, J. Kim, P. Gonzalez-Calero, P. Groth, J. Moody, and E. Deelman. 2011. “Wings: Intelligent Workflow-Based Design of Computational Experiments.” <em>IEEE Intelligent Systems</em> 26 (1): 62–72. doi:<a href="https://doi.org/10.1109/MIS.2010.9">10.1109/MIS.2010.9</a>.</p>
</div>
<div id="ref-PROVModel">
<p>Gil, Yolanda, Simon Miles, Khalid Belhajjame, Helena Deus, Daniel Garijo, Graham Klyne, Paolo Missier, Stian Soiland-Reyes, and Stephan Zednik. 2013. “PROV Model Primer.” April 30. <a href="http://www.w3.org/TR/2013/NOTE-prov-primer-20130430/" class="uri">http://www.w3.org/TR/2013/NOTE-prov-primer-20130430/</a>.</p>
</div>
<div id="ref-GitHub">
<p>“GitHub.” <a href="https://github.com/" class="uri">https://github.com/</a>.</p>
</div>
<div id="ref-Goble2014">
<p>Goble, Carole. 2014. “Better Software, Better Research.” <em>IEEE Internet Computing</em> 18 (5). Institute of Electrical; Electronics Engineers (IEEE): 4–8. doi:<a href="https://doi.org/10.1109/mic.2014.88">10.1109/mic.2014.88</a>.</p>
</div>
<div id="ref-Goble2010">
<p>Goble, Carole A., Jiten Bhagat, Sergejs Aleksejevs, Don Cruickshank, Danius Michaelides, David Newman, Mark Borkum, et al. 2010. “myExperiment: A Repository and Social Network for the Sharing of Bioinformatics Workflows.” <em>Nucleic Acids Research</em> 38 (suppl_2). Oxford University Press (OUP): W677–W682. doi:<a href="https://doi.org/10.1093/nar/gkq429">10.1093/nar/gkq429</a>.</p>
</div>
<div id="ref-gomezperez_2017">
<p>Gomez-Perez, Jose Manuel, Raul Palma, and Andres Garcia-Silva. 2017. “Towards a Human-Machine Scientific Partnership Based on Semantically Rich Research Objects.” In <em>2017 IEEE 13th International Conference on E-Science (E-Science)</em>, 266–75. IEEE. doi:<a href="https://doi.org/10.1109/eScience.2017.40">10.1109/eScience.2017.40</a>.</p>
</div>
<div id="ref-GoogleCl">
<p>“Google Cloud Including Gcp &amp; G Suite Url = https://cloud.google.com/, Note = Accessed 01 Dec 2018, Urldate = 01/12/2018.” n.d.</p>
</div>
<div id="ref-Gruening2018">
<p>Gruening, Bjorn, Olivier Sallou, Pablo Moreno, Felipe da Veiga Leprevost, Hervé Ménager, Dan Søndergaard, Hannes Röst, et al. 2018. “Recommendations for the Packaging and Containerizing of Bioinformatics Software.” <em>F1000Research</em> 7 (June). F1000 Research, Ltd.: 742. doi:<a href="https://doi.org/10.12688/f1000research.15140.1">10.12688/f1000research.15140.1</a>.</p>
</div>
<div id="ref-Grning2018">
<p>Grüning, Björn, and Ryan Dale, Andreas Sjödin, Brad A. Chapman, Jillian Rowe, Christopher H. Tomkins-Tinch, Renan Valieris, and Johannes Köster. 2018. “Bioconda: Sustainable and Comprehensive Software Distribution for the Life Sciences.” <em>Nature Methods</em> 15 (7). Springer Nature: 475–76. doi:<a href="https://doi.org/10.1038/s41592-018-0046-7">10.1038/s41592-018-0046-7</a>.</p>
</div>
<div id="ref-gtexpipe">
<p>“Gtex Rna-Seq Pipeline.” 2017. <a href="https://github.com/broadinstitute/gtex-pipeline/tree/master/rnaseq" class="uri">https://github.com/broadinstitute/gtex-pipeline/tree/master/rnaseq</a>.</p>
</div>
<div id="ref-guimera_2012">
<p>Guimera, Roman Valls. 2012. “Bcbio-Nextgen: Automated, Distributed Next-Gen Sequencing Pipeline.” <em>EMBnet J.</em> 17 (B): 30. doi:<a href="https://doi.org/10.14806/ej.17.B.286">10.14806/ej.17.B.286</a>.</p>
</div>
<div id="ref-Gymrek2016">
<p>Gymrek, Melissa, and Yossi Farjoun. 2016. “Recommendations for Open Data Science.” <em>GigaScience</em> 5 (1). Oxford University Press (OUP). doi:<a href="https://doi.org/10.1186/s13742-016-0127-4">10.1186/s13742-016-0127-4</a>.</p>
</div>
<div id="ref-heliumda">
<p>“Heliumdatacommons.” 2017. <a href="https://github.com/heliumdatacommons" class="uri">https://github.com/heliumdatacommons</a>.</p>
</div>
<div id="ref-NIH-PILOT">
<p>“Heliumdatacommons/Cwl_workflows: Example Cwl Workflows That Run on Team Helium Pivot Architecture.” 11AD–10AD. <a href="https://github.com/heliumdatacommons/cwl_workflows" class="uri">https://github.com/heliumdatacommons/cwl_workflows</a>.</p>
</div>
<div id="ref-herschel_2017">
<p>Herschel, Melanie, Ralf Diestelkämper, and Houssem Ben Lahmar. 2017. “A Survey on Provenance: What for? What Form? What from?” <em>VLDB J.</em> 26 (6): 881–906.</p>
</div>
<div id="ref-hettne_2014">
<p>Hettne, Kristina M, Harish Dharuri, Jun Zhao, Katherine Wolstencroft, Khalid Belhajjame, Stian Soiland-Reyes, Eleni Mina, et al. 2014. “Structuring Research Methods and Data with the Research Object Model: Genomics Workflows as a Case Study.” <em>J Biomed Semantics</em> 5 (1): 41. doi:<a href="https://doi.org/10.1186/2041-1480-5-41">10.1186/2041-1480-5-41</a>.</p>
</div>
<div id="ref-howe">
<p>Howe, Bill. 2012. “Virtual Appliances, Cloud Computing, and Reproducible Research.” <em>Computing in Science &amp; Engineering</em> 14 (4). Institute of Electrical; Electronics Engineers (IEEE): 36–41. doi:<a href="https://doi.org/10.1109/mcse.2012.62">10.1109/mcse.2012.62</a>.</p>
</div>
<div id="ref-PROVXML">
<p>Hua, Hook, Curt Tilmes, Stephan Zednik, and Luc Moreau. 2013. “PROV-Xml: The Prov Xml Schema.” April 30. <a href="http://www.w3.org/TR/2013/NOTE-prov-xml-20130430/" class="uri">http://www.w3.org/TR/2013/NOTE-prov-xml-20130430/</a>.</p>
</div>
<div id="ref-huynh_2013">
<p>Huynh, Trung Dong, Michael Jewell, Amir Sezavar Keshavarz, Danius Michaelides, Huanjia Yang, and Luc Moreau. 2013. “The PROV-JSON Serialization. a JSON Representation for the PROV Data Model.” April 24. <a href="http://www.w3.org/Submission/2013/SUBM-prov-json-20130424/" class="uri">http://www.w3.org/Submission/2013/SUBM-prov-json-20130424/</a>.</p>
</div>
<div id="ref-interope55:online">
<p>“Interoperability | Definition of Interoperability in English by Oxford Dictionaries.” <a href="https://en.oxforddictionaries.com/definition/interoperability" class="uri">https://en.oxforddictionaries.com/definition/interoperability</a>.</p>
</div>
<div id="ref-variantcallinglecture">
<p>“Introduction to Variant Calling.” 2014. <a href="https://bioconductor.org/help/course-materials/2014/CSAMA2014/3_Wednesday/lectures/VariantCallingLecture.pdf" class="uri">https://bioconductor.org/help/course-materials/2014/CSAMA2014/3_Wednesday/lectures/VariantCallingLecture.pdf</a>.</p>
</div>
<div id="ref-Ison2013">
<p>Ison, J., M. Kalas, I. Jonassen, D. Bolser, M. Uludag, H. McWilliam, J. Malone, R. Lopez, S. Pettifer, and P. Rice. 2013. “EDAM: An Ontology of Bioinformatics Operations, Types of Data and Identifiers, Topics and Formats.” <em>Bioinformatics</em> 29 (10). Oxford University Press (OUP): 1325–32. doi:<a href="https://doi.org/10.1093/bioinformatics/btt113">10.1093/bioinformatics/btt113</a>.</p>
</div>
<div id="ref-Ivie2018">
<p>Ivie, Peter, and Douglas Thain. 2018. “Reproducibility in Scientific Computing.” <em>ACM Comput. Surv.</em> 51 (3). New York, NY, USA: ACM: 63:1–63:36. doi:<a href="https://doi.org/10.1145/3186266">10.1145/3186266</a>.</p>
</div>
<div id="ref-kanwal_2017">
<p>Kanwal, Sehrish, Farah Zaib Khan, Andrew Lonie, and Richard O Sinnott. 2017. “Investigating Reproducibility and Tracking Provenance - a Genomic Workflow Case Study.” <em>BMC Bioinformatics</em> 18 (1): 337. doi:<a href="https://doi.org/10.1186/s12859-017-1747-0">10.1186/s12859-017-1747-0</a>.</p>
</div>
<div id="ref-kanwal2017digital">
<p>Kanwal, Sehrish, Andrew Lonie, and Richard O Sinnott. 2017. “Digital Reproducibility Requirements of Computational Genomic Workflows.” In <em>Bioinformatics and Biomedicine (Bibm), 2017 Ieee International Conference on</em>, 1522–9. IEEE. doi:<a href="https://doi.org/10.1109/bibm.2017.8217887">10.1109/bibm.2017.8217887</a>.</p>
</div>
<div id="ref-kaushik_2017">
<p>Kaushik, Gaurav, Sinisa Ivkovic, Janko Simonovic, Nebojsa Tijanic, Brandi Davis-Dusenbery, and Deniz Kural. 2017. “Rabix: An Open-Source Workflow Executor Supporting Recomputability and Interoperability of Workflow Descriptions.” <em>Pac Symp Biocomput</em> 22: 154–65. doi:<a href="https://doi.org/10.1142/9789813207813\_0016">10.1142/9789813207813\_0016</a>.</p>
</div>
<div id="ref-alignment_mendeley">
<p>Khan, Farah Zaib, and Stian Soiland-Reyes. 2018a. “CWL Run of Alignment Workflow (Cwlprov 0.6.0 Research Object).” Mendeley Data. doi:<a href="https://doi.org/10.17632/6wtpgr3kbj.1">10.17632/6wtpgr3kbj.1</a>.</p>
</div>
<div id="ref-rnaseq_mendeley">
<p>———. 2018b. “CWL Run of Rna-Seq Analysis Workflow (Cwlprov 0.5.0 Research Object)).” Mendeley Data. doi:<a href="https://doi.org/10.17632/xnwncxpw42.1">10.17632/xnwncxpw42.1</a>.</p>
</div>
<div id="ref-somatic_mendeley">
<p>———. 2018c. “CWL Run of Somatic Variant Calling Workflow (Cwlprov 0.5.0 Research Object).” Mendeley Data. doi:<a href="https://doi.org/10.17632/97hj93mkfd.3">10.17632/97hj93mkfd.3</a>.</p>
</div>
<div id="ref-cwlprov-preprint">
<p>Khan, Farah Zaib, Stian Soiland-Reyes, Michael R. Crusoe, Andrew Lonie, and Richard Sinnott. 2018. “CWLProv - Interoperable Retrospective Provenance Capture and Its Challenges,” March. Zenodo. doi:<a href="https://doi.org/10.5281/zenodo.1215611">10.5281/zenodo.1215611</a>.</p>
</div>
<div id="ref-kim2016assessing">
<p>Kim, Donghoon, and Mladen A Vouk. 2016. “Assessing Run-Time Overhead of Securing Kepler.” <em>Procedia Computer Science</em> 80. Elsevier: 2281–6. doi:<a href="https://doi.org/10.1016/j.procs.2016.05.412">10.1016/j.procs.2016.05.412</a>.</p>
</div>
<div id="ref-5star">
<p>Kim, James G., and Michael Hausenblas et al. 21AD–5AD. “5 Star Open Data.” <a href="https://5stardata.info/" class="uri">https://5stardata.info/</a>.</p>
</div>
<div id="ref-cwlairflow2018">
<p>Kotliar, Michael, Andrey Kartashov, and Artem Barski. 2018. “CWL-Airflow: A Lightweight Pipeline Manager Supporting Common Workflow Language.” <em>bioRxiv</em>. Cold Spring Harbor Laboratory. doi:<a href="https://doi.org/10.1101/249243">10.1101/249243</a>.</p>
</div>
<div id="ref-kurtzer_2017">
<p>Kurtzer, Gregory M, Vanessa Sochat, and Michael W Bauer. 2017. “Singularity: Scientific Containers for Mobility of Compute.” <em>PLoS ONE</em> 12 (5): e0177459. doi:<a href="https://doi.org/10.1371/journal.pone.0177459">10.1371/journal.pone.0177459</a>.</p>
</div>
<div id="ref-PROVO">
<p>Lebo, Timothy, Satya Sahoo, Deborah McGuinness, Khalid Belhajjame, James Cheney, David Corsar, Daniel Garijo, Stian Soiland-Reyes, Stephan Zednik, and Jun Zhao. 2013. “PROV-O: The Prov Ontology.” April 30. <a href="http://www.w3.org/TR/2013/REC-prov-o-20130430/" class="uri">http://www.w3.org/TR/2013/REC-prov-o-20130430/</a>.</p>
</div>
<div id="ref-li2011rsem">
<p>Li, Bo, and Colin N Dewey. 2011. “RSEM: Accurate Transcript Quantification from Rna-Seq Data with or Without a Reference Genome.” <em>BMC Bioinformatics</em> 12 (1). BioMed Central: 323. doi:<a href="https://doi.org/10.1186/1471-2105-12-323">10.1186/1471-2105-12-323</a>.</p>
</div>
<div id="ref-li2013aligning">
<p>Li, Heng. 2013. “Aligning Sequence Reads, Clone Sequences and Assembly Contigs with Bwa-Mem.” <em>arXiv Preprint arXiv:1303.3997</em>. <a href="https://arxiv.org/abs/1303.3997" class="uri">https://arxiv.org/abs/1303.3997</a>.</p>
</div>
<div id="ref-li2009sequence">
<p>Li, Heng, Bob Handsaker, Alec Wysoker, Tim Fennell, Jue Ruan, Nils Homer, Gabor Marth, Goncalo Abecasis, and Richard Durbin. 2009. “The Sequence Alignment/Map Format and Samtools.” <em>Bioinformatics</em> 25 (16). Oxford University Press: 2078–9. doi:<a href="https://doi.org/10.1093/bioinformatics/btp352">10.1093/bioinformatics/btp352</a>.</p>
</div>
<div id="ref-Littauer2012">
<p>Littauer, Richard, Karthik Ram, Bertram Ludäscher, William Michener, and Rebecca Koskela. 2012. “Trends in Use of Scientific Workflows: Insights from a Public Repository and Recommendations for Best Practice.” <em>International Journal of Digital Curation</em> 7 (2). Edinburgh University Library: 92–100. doi:<a href="https://doi.org/10.2218/ijdc.v7i2.232">10.2218/ijdc.v7i2.232</a>.</p>
</div>
<div id="ref-reusevocab">
<p>Lóscio, Bernadette Farias, Caroline Burle, Newton Calegari, Annette Greiner Antoine Isaac, Carlos Iglesias, Carlos Laufer, Christophe Guéret, et al. 2017. “Data on the Web Best Practices.” January 31. <a href="https://www.w3.org/TR/2017/REC-dwbp-20170131/#ReuseVocabularies" class="uri">https://www.w3.org/TR/2017/REC-dwbp-20170131/#ReuseVocabularies</a>.</p>
</div>
<div id="ref-ludascher2016brief">
<p>Ludäscher, Bertram. 2016. “A Brief Tour Through Provenance in Scientific Workflows and Databases.” <em>Springer Proceedings in Business and Economics</em>. Springer International Publishing, 103–26. doi:<a href="https://doi.org/10.1007/978-3-319-40226-0_7">10.1007/978-3-319-40226-0_7</a>.</p>
</div>
<div id="ref-madduri_2018">
<p>Madduri, Kyle AND D’Arcy, Ravi AND Chard. 2019. “Reproducible Big Data Science: A Case Study in Continuous Fairness.” <em>PLOS ONE</em> 14 (4). Public Library of Science: 1–22. doi:<a href="https://doi.org/10.1371/journal.pone.0213013">10.1371/journal.pone.0213013</a>.</p>
</div>
<div id="ref-Malawski_2013">
<p>Malawski, Maciej, Kamil Figiela, and Jarek Nabrzyski. 2013. “Cost Minimization for Computational Applications on Hybrid Cloud Infrastructures.” <em>Future Generation Computer Systems</em> 29 (7). Elsevier BV: 1786–94. doi:<a href="https://doi.org/10.1016/j.future.2013.01.004">10.1016/j.future.2013.01.004</a>.</p>
</div>
<div id="ref-docker">
<p>Merkel, Dirk. 2014. “Docker: Lightweight Linux Containers for Consistent Development and Deployment.” <em>Linux Journal</em> 2014 (239). Houston, TX: Belltown Media. <a href="https://www.linuxjournal.com/node/1335702" class="uri">https://www.linuxjournal.com/node/1335702</a>.</p>
</div>
<div id="ref-michaelides_2016">
<p>Michaelides, Danius T., Richard Parker, Chris Charlton, William J. Browne, and Luc Moreau. 2016. “Intermediate Notation for Provenance and Workflow Reproducibility.” In <em>Provenance and Annotation of Data and Processes</em>, edited by Marta Mattoso and Boris Glavic, 9672:83–94. Lecture Notes in Computer Science. Cham: Springer International Publishing. doi:<a href="https://doi.org/10.1007/978-3-319-40593-3\_7">10.1007/978-3-319-40593-3\_7</a>.</p>
</div>
<div id="ref-Michel_2018">
<p>Michel, Franck, and The Bioschemas Community. 2018. “Bioschemas &amp;Amp; Schema.org: A Lightweight Semantic Layer for Life Sciences Websites.” <em>Biodiversity Information Science and Standards</em> 2 (May). Pensoft Publishers: e25836. doi:<a href="https://doi.org/10.3897/biss.2.25836">10.3897/biss.2.25836</a>.</p>
</div>
<div id="ref-Microsof">
<p>“Microsoft Azure Cloud Computing Platform &amp; Services.” 1AD–12AD. <a href="https://azure.microsoft.com/en-us/" class="uri">https://azure.microsoft.com/en-us/</a>.</p>
</div>
<div id="ref-Missier2016">
<p>Missier, Paolo. 2016. “The Lifecycle of Provenance Metadata and Its Associated Challenges and Opportunities.” In <em>Building Trust in Information</em>, 127–37. Springer International Publishing. doi:<a href="https://doi.org/10.1007/978-3-319-40226-0_8">10.1007/978-3-319-40226-0_8</a>.</p>
</div>
<div id="ref-Missier2013">
<p>Missier, Paolo, Khalid Belhajjame, and James Cheney. 2013. “The W3c Prov Family of Specifications for Modelling Provenance Metadata.” In <em>Proceedings of the 16th International Conference on Extending Database Technology</em>, 773–76. EDBT ’13. New York, NY, USA: ACM. doi:<a href="https://doi.org/10.1145/2452376.2452478">10.1145/2452376.2452478</a>.</p>
</div>
<div id="ref-mitchell_2017">
<p>Mitchell, Alex L, Maxim Scheremetjew, Hubert Denise, Simon Potter, Aleksandra Tarkowska, Matloob Qureshi, Gustavo A Salazar, et al. 2018. “EBI Metagenomics in 2017: Enriching the Analysis of Microbial Communities, from Sequence Reads to Assemblies.” <em>Nucleic Acids Research</em> 46 (D1): D726–D735. doi:<a href="https://doi.org/10.1093/nar/gkx967">10.1093/nar/gkx967</a>.</p>
</div>
<div id="ref-Mohan2014">
<p>Mohan, Aravind, Shiyong Lu, and Alexander Kotov. 2014. “Addressing the Shimming Problem in Big Data Scientific Workflows.” In <em>2014 IEEE International Conference on Services Computing</em>. IEEE. doi:<a href="https://doi.org/10.1109/scc.2014.53">10.1109/scc.2014.53</a>.</p>
</div>
<div id="ref-debian-med">
<p>Möller, Steffen, Stuart W. Prescott, Lars Wirzenius, Petter Reinholdtsen, Brad Chapman, Pjotr Prins, Stian Soiland-Reyes, et al. 2017. “Robust Cross-Platform Workflows: How Technical and Scientific Communities Collaborate to Develop, Test and Share Best Practices for Data Analysis.” <em>Data Science and Engineering</em> 2 (3): 232–44. doi:<a href="https://doi.org/10.1007/s41019-017-0050-4">10.1007/s41019-017-0050-4</a>.</p>
</div>
<div id="ref-moller_2017">
<p>Möller, Steffen, Stuart W. Prescott, Lars Wirzenius, Petter Reinholdtsen, Brad Chapman, Pjotr Prins, Stian Soiland-Reyes, et al. 2017. “Robust Cross-Platform Workflows: How Technical and Scientific Communities Collaborate to Develop, Test and Share Best Practices for Data Analysis.” <em>Data Sci. Eng.</em> 2 (3): 232–44. doi:<a href="https://doi.org/10.1007/s41019-017-0050-4">10.1007/s41019-017-0050-4</a>.</p>
</div>
<div id="ref-Moreau2008">
<p>Moreau, Luc, Juliana Freire, Joe Futrelle, Robert E. McGrath, Jim Myers, and Patrick Paulson. 2008. “The Open Provenance Model: An Overview.” In <em>Lecture Notes in Computer Science</em>, 323–26. Springer Berlin Heidelberg. doi:<a href="https://doi.org/10.1007/978-3-540-89965-5_31">10.1007/978-3-540-89965-5_31</a>.</p>
</div>
<div id="ref-moreau2009governance">
<p>Moreau, Luc, Juliana Freire, Joe Futrelle, Jim Myers, and Patrick Paulson. 2009. “Governance of the Open Provenance Model,” June. <a href="https://nms.kcl.ac.uk/luc.moreau/papers/governance.pdf" class="uri">https://nms.kcl.ac.uk/luc.moreau/papers/governance.pdf</a>.</p>
</div>
<div id="ref-Moreau2015">
<p>Moreau, Luc, Paul Groth, James Cheney, Timothy Lebo, and Simon Miles. 2015. “The Rationale of PROV.” <em>Web Semantics: Science, Services and Agents on the World Wide Web</em> 35 (December). Elsevier BV: 235–57. doi:<a href="https://doi.org/10.1016/j.websem.2015.04.001">10.1016/j.websem.2015.04.001</a>.</p>
</div>
<div id="ref-PROVDM">
<p>Moreau, Luc, Paolo Missier, Khalid Belhajjame, Reza B’Far, James Cheney, Sam Coppens, Stephen Cresswell, et al. 2013. “PROV-Dm: The PROV Data Model.” W3C Recommendation. W3C. April.</p>
</div>
<div id="ref-moreau_2013">
<p>Moreau, Luc, Paolo Missier, James Cheney, and Stian Soiland-Reyes. 2013. “PROV-N: The Provenance Notation.” April 30. <a href="http://www.w3.org/TR/2013/REC-prov-n-20130430/" class="uri">http://www.w3.org/TR/2013/REC-prov-n-20130430/</a>.</p>
</div>
<div id="ref-Mozilla">
<p>“Mozilla Science.” 1AD–12AD. <a href="https://science.mozilla.org/" class="uri">https://science.mozilla.org/</a>.</p>
</div>
<div id="ref-Nectar">
<p>“Nectar Cloud - Nectar.” <a href="https://nectar.org.au/research-cloud/" class="uri">https://nectar.org.au/research-cloud/</a>.</p>
</div>
<div id="ref-Nekrutenko2012">
<p>Nekrutenko, Anton, and James Taylor. 2012. “Next-Generation Sequencing Data Interpretation: Enhancing Reproducibility and Accessibility.” <em>Nature Reviews Genetics</em> 13 (9). Springer Nature: 667–72. doi:<a href="https://doi.org/10.1038/nrg3305">10.1038/nrg3305</a>.</p>
</div>
<div id="ref-bagit17">
<p>Network Working Group. 2017. “Draft-Kunze-Bagit-17 - the Bagit File Packaging Format (V1.0).” <a href="https://tools.ietf.org/html/draft-kunze-bagit-17" class="uri">https://tools.ietf.org/html/draft-kunze-bagit-17</a>.</p>
</div>
<div id="ref-OBFTravel">
<p>“OBF Travel Fellowship Program | Obf News.” 2016. <a href="https://news.open-bio.org/2016/03/01/obf-travel-fellowship-program/" class="uri">https://news.open-bio.org/2016/03/01/obf-travel-fellowship-program/</a>.</p>
</div>
<div id="ref-tazro2018">
<p>Ohta, Tazro, Tomoya Tanjo, and Osamu Ogasawara. 2018. “Accumulating Computational Resource Usage of Genomic Data Analysis Workflow to Optimize Cloud Computing Instance Selection.” <em>bioRxiv</em>. doi:<a href="https://doi.org/10.1101/456756">10.1101/456756</a>.</p>
</div>
<div id="ref-10.1093/gigascience/giz052">
<p>———. 2019. “Accumulating computational resource usage of genomic data analysis workflow to optimize cloud computing instance selection.” <em>GigaScience</em> 8 (4). doi:<a href="https://doi.org/10.1093/gigascience/giz052">10.1093/gigascience/giz052</a>.</p>
</div>
<div id="ref-pasquier_2017">
<p>Pasquier, Thomas, Xueyuan Han, Mark Goldstein, Thomas Moyer, David Eyers, Margo Seltzer, and Jean Bacon. 2017. “Practical Whole-System Provenance Capture.” In <em>Proceedings of the 2017 Symposium on Cloud Computing - SoCC ’17</em>, 405–18. New York, New York, USA: ACM Press. doi:<a href="https://doi.org/10.1145/3127479.3129249">10.1145/3127479.3129249</a>.</p>
</div>
<div id="ref-provpython">
<p>“Prov 1.5.2.” 2018. <a href="https://pypi.org/project/prov/" class="uri">https://pypi.org/project/prov/</a>.</p>
</div>
<div id="ref-roprinciples">
<p>“Researchobject.org.” <a href="http://www.researchobject.org/overview/" class="uri">http://www.researchobject.org/overview/</a>.</p>
</div>
<div id="ref-sandve_2013">
<p>Sandve, Geir Kjetil, Anton Nekrutenko, James Taylor, and Eivind Hovig. 2013. “Ten Simple Rules for Reproducible Computational Research.” <em>PLoS Comput Biol</em> 9 (10): e1003285. doi:<a href="https://doi.org/10.1371/journal.pcbi.1003285">10.1371/journal.pcbi.1003285</a>.</p>
</div>
<div id="ref-Saunders2012">
<p>Saunders, Christopher T., Wendy S. W. Wong, Sajani Swamy, Jennifer Becq, Lisa J. Murray, and R. Keira Cheetham. 2012. “Strelka: Accurate Somatic Small-Variant Calling from Sequenced Tumornormal Sample Pairs.” <em>Bioinformatics</em> 28 (14). Oxford University Press (OUP): 1811–7. doi:<a href="https://doi.org/10.1093/bioinformatics/bts271">10.1093/bioinformatics/bts271</a>.</p>
</div>
<div id="ref-datacrate_2018">
<p>Sefton, Peter, Michael Lynch, Gerard Devine, and Duncan Loxton. 2018. “DataCrate: A Method of Packaging, Distributing, Displaying and Archiving Research Objects,” July. doi:<a href="https://doi.org/10.5281/zenodo.1312323">10.5281/zenodo.1312323</a>.</p>
</div>
<div id="ref-seo2012transcriptional">
<p>Seo, Jeong-Sun, Young Seok Ju, Won-Chul Lee, Jong-Yeon Shin, June Koo Lee, Thomas Bleazard, Junho Lee, et al. 2012. “The Transcriptional Landscape and Mutational Profile of Lung Adenocarcinoma.” <em>Genome Research</em>. Cold Spring Harbor Lab. doi:<a href="https://doi.org/10.1101/gr.145144.112">10.1101/gr.145144.112</a>.</p>
</div>
<div id="ref-softwarecarpentry">
<p>“Software Carpentry.” 4AD–10AD. <a href="https://software-carpentry.org/" class="uri">https://software-carpentry.org/</a>.</p>
</div>
<div id="ref-tavernaprov">
<p>Soiland-Reyes, Stian, Pinar Alper, and Carole Goble. 2016. “Tracking Workflow Execution with TavernaProv.” June. doi:<a href="https://doi.org/10.5281/zenodo.51314">10.5281/zenodo.51314</a>.</p>
</div>
<div id="ref-soilandreyes_2018">
<p>Soiland-Reyes, Stian, and Marcos Cáceres. 2018a. “The Archive and Package (Arcp) URI Scheme.” <em>Internet-Draft Soilandreyes-Arcp</em>, January. <a href="https://tools.ietf.org/id/draft-soilandreyes-arcp-03.html" class="uri">https://tools.ietf.org/id/draft-soilandreyes-arcp-03.html</a>.</p>
</div>
<div id="ref-arcp_ro2018">
<p>———. 2018b. “The Archive and Package (Arcp) Uri Scheme.” In <em>2018 Ieee 13th International Conference on E-Science (E-Science)</em>. <a href="http://s11.no/2018/arcp.html" class="uri">http://s11.no/2018/arcp.html</a>.</p>
</div>
<div id="ref-cwlprov-py">
<p>Soiland-Reyes, Stian, and Farah Zaib Khan. 2018. “common-workflow-language/cwlprov-py: cwlprov-py 0.1.1.” Zenodo. doi:<a href="https://doi.org/10.5281/zenodo.1471376">10.5281/zenodo.1471376</a>.</p>
</div>
<div id="ref-wf4ever2">
<p>Soiland-Reyes, Stian, Sean Bechhofer, Oscar Corcho, Graham Klyne, Khalid Belhajjame, Daniel Garijo, Esteban García Cuesta, and Raul Palma. 2016a. “The Wfprov Ontology.” <a href="https://w3id.org/ro/2016-01-28/wfprov/" class="uri">https://w3id.org/ro/2016-01-28/wfprov/</a>.</p>
</div>
<div id="ref-wf4ever1">
<p>———. 2016b. “The Wfdesc Ontology.” January 28. <a href="https://w3id.org/ro/2016-01-28/wfdesc" class="uri">https://w3id.org/ro/2016-01-28/wfdesc</a>.</p>
</div>
<div id="ref-RObundle">
<p>Soiland-Reyes, Stian, Matthew Gamble, and Robert Haines. 2014. “Research Object Bundle 1.0.” Zenodo. doi:<a href="https://doi.org/10.5281/zenodo.12586">10.5281/zenodo.12586</a>.</p>
</div>
<div id="ref-cwlprov">
<p>Soiland-Reyes, Stian, Farah Zaib Khan, and Michael R Crusoe. 2018. “Common-Workflow-Language/Cwlprov: CWLProv 0.6.0.” Zenodo. doi:<a href="https://doi.org/10.5281/zenodo.1471585">10.5281/zenodo.1471585</a>.</p>
</div>
<div id="ref-bcbiowf">
<p>“Somatic Variant Calling Workflow.” 2018. <a href="https://github.com/FarahZKhan/bcbio_test_cwlprov/blob/master/somatic/somatic-workflow/main-somatic.cwl" class="uri">https://github.com/FarahZKhan/bcbio_test_cwlprov/blob/master/somatic/somatic-workflow/main-somatic.cwl</a>.</p>
</div>
<div id="ref-Spjuth2015">
<p>Spjuth, Ola, Erik Bongcam-Rudloff, Guillermo Carrasco Hernández, Lukas Forer, Mario Giovacchini, Roman Valls Guimera, Aleksi Kallio, et al. 2015. “Experiences with Workflows for Automating Data-Intensive Bioinformatics.” <em>Biology Direct</em> 10 (1). Springer Nature. doi:<a href="https://doi.org/10.1186/s13062-015-0071-8">10.1186/s13062-015-0071-8</a>.</p>
</div>
<div id="ref-JSONLD">
<p>Sporny, Manu, Dave Longley, Gregg Kellogg, Markus Lanthaler, and Niklas Lindström. 2014. “JSON-Ld 1.0: A Json-Based Serialization for Linked Data.” January 16. <a href="http://www.w3.org/TR/2014/REC-json-ld-20140116/" class="uri">http://www.w3.org/TR/2014/REC-json-ld-20140116/</a>.</p>
</div>
<div id="ref-docker-alignment">
<p>“Statgen/Docker-Alignment: Dockerfile for Alignment.” 2017. <a href="https://github.com/statgen/docker-alignment" class="uri">https://github.com/statgen/docker-alignment</a>.</p>
</div>
<div id="ref-stephens_2015">
<p>Stephens, Zachary D, Skylar Y Lee, Faraz Faghri, Roy H Campbell, Chengxiang Zhai, Miles J Efron, Ravishankar Iyer, Michael C Schatz, Saurabh Sinha, and Gene E Robinson. 2015. “Big Data: Astronomical or Genomical?” <em>PLoS Biol</em> 13 (7): e1002195. doi:<a href="https://doi.org/10.1371/journal.pbio.1002195">10.1371/journal.pbio.1002195</a>.</p>
</div>
<div id="ref-Stodden2014">
<p>Stodden, Victoria, and Sheila Miguez. 2014. “Best Practices for Computational Science: Software Infrastructure and Environments for Reproducible and Extensible Research.” <em>Journal of Open Research Software</em> 2 (1). Ubiquity Press, Ltd. doi:<a href="https://doi.org/10.5334/jors.ay">10.5334/jors.ay</a>.</p>
</div>
<div id="ref-stodden_2016">
<p>Stodden, Victoria, Marcia McNutt, David H Bailey, Ewa Deelman, Yolanda Gil, Brooks Hanson, Michael A Heroux, John P A Ioannidis, and Michela Taufer. 2016. “Enhancing Reproducibility for Computational Methods.” <em>Science</em> 354 (6317): 1240–1. doi:<a href="https://doi.org/10.1126/science.aah6168">10.1126/science.aah6168</a>.</p>
</div>
<div id="ref-Tolk">
<p>Tolk, A. n.d. “What Comes After the Semantic Web - PADS Implications for the Dynamic Web.” In <em>20th Workshop on Principles of Advanced and Distributed Simulation (PADS06)</em>. IEEE. doi:<a href="https://doi.org/10.1109/pads.2006.39">10.1109/pads.2006.39</a>.</p>
</div>
<div id="ref-MarkDuplicates">
<p>“Tool Documentation: MarkDuplicates.” <a href="http://broadinstitute.github.io/picard/command-line-overview.html#MarkDuplicates" class="uri">http://broadinstitute.github.io/picard/command-line-overview.html#MarkDuplicates</a>.</p>
</div>
<div id="ref-rnaseq">
<p>“TopMed Rna-Seq Workflow.” 2018. <a href="https://w3id.org/cwl/view/git/027e8af41b906173aafdb791351fb29efc044120/topmed-workflows/TOPMed_RNAseq_pipeline/rnaseq_pipeline_fastq.cwl" class="uri">https://w3id.org/cwl/view/git/027e8af41b906173aafdb791351fb29efc044120/topmed-workflows/TOPMed_RNAseq_pipeline/rnaseq_pipeline_fastq.cwl</a>.</p>
</div>
<div id="ref-alignment-wf">
<p>“Topmed-Workflows/Topmed-Alignment.cwl at Cwlprov_testing · Farahzkhan/Topmed-Workflows.” <a href="https://github.com/FarahZKhan/topmed-workflows/blob/2fcea0b9469b572399755b6828ff87a40d865e43/aligner/sbg-alignment-cwl/topmed-alignment.cwl" class="uri">https://github.com/FarahZKhan/topmed-workflows/blob/2fcea0b9469b572399755b6828ff87a40d865e43/aligner/sbg-alignment-cwl/topmed-alignment.cwl</a>.</p>
</div>
<div id="ref-TransOmi24">
<p>“Trans-Omics for Precision Medicine (Topmed) Program | National Heart, Lung, and Blood Institute (Nhlbi).” 2014. <a href="https://www.nhlbi.nih.gov/science/trans-omics-precision-medicine-topmed-program" class="uri">https://www.nhlbi.nih.gov/science/trans-omics-precision-medicine-topmed-program</a>.</p>
</div>
<div id="ref-vivian2017toil">
<p>Vivian, John, Arjun Arkal Rao, Frank Austin Nothaft, Christopher Ketchum, Joel Armstrong, Adam Novak, Jacob Pfeil, et al. 2017. “Toil Enables Reproducible, Open Source, Big Biomedical Data Analyses.” <em>Nature Biotechnology</em> 35 (4). Nature Publishing Group: 314. doi:<a href="https://doi.org/10.1038/nbt.3772">10.1038/nbt.3772</a>.</p>
</div>
<div id="ref-cromwell">
<p>Voss, Kate, Geraldine Van Der Auwera, and Jeff Gentry. 2017. “Full-Stack Genomics Pipelining with Gatk4 + Wdl + Cromwell.” F1000Research. doi:<a href="https://doi.org/10.7490/f1000research.1114634.1">10.7490/f1000research.1114634.1</a>.</p>
</div>
<div id="ref-W3CProvWorkingGroup">
<p>“W3C Provenance Incubator Group Wiki - Xg Provenance Wiki.” <a href="https://www.w3.org/2005/Incubator/prov/wiki/W3C\_Provenance\_Incubator\_Group\_Wiki" class="uri">https://www.w3.org/2005/Incubator/prov/wiki/W3C\_Provenance\_Incubator\_Group\_Wiki</a>.</p>
</div>
<div id="ref-wilkinson_2016">
<p>Wilkinson, Mark D, Michel Dumontier, I Jsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, et al. 2016. “The FAIR Guiding Principles for Scientific Data Management and Stewardship.” <em>Sci Data</em> 3 (March): 160018. doi:<a href="https://doi.org/10.1038/sdata.2016.18">10.1038/sdata.2016.18</a>.</p>
</div>
<div id="ref-wolstencroft_2013">
<p>Wolstencroft, Katherine, Robert Haines, Donal Fellows, Alan Williams, David Withers, Stuart Owen, Stian Soiland-Reyes, et al. 2013. “The Taverna Workflow Suite: Designing and Executing Workflows of Web Services on the Desktop, Web or in the Cloud.” <em>Nucleic Acids Res</em> 41 (Web Server issue): W557–61. doi:<a href="https://doi.org/10.1093/nar/gkt328">10.1093/nar/gkt328</a>.</p>
</div>
<div id="ref-biosimspacewebinar">
<p>Woods, Christopher. 2018. “BioExcel Webinar #28: BioSimSpace – Filling the Gaps Between Molecular Simulation Codes.” YouTube, BioExcel. June. <a href="https://youtu.be/pD8mhj3WElE?t=1599" class="uri">https://youtu.be/pD8mhj3WElE?t=1599</a>.</p>
</div>
<div id="ref-xu2018review">
<p>Xu, Chang. 2018. “A Review of Somatic Single Nucleotide Variant Calling Algorithms for Next-Generation Sequencing Data.” <em>Computational and Structural Biotechnology Journal</em>. Elsevier. doi:<a href="https://doi.org/10.1016/j.csbj.2018.01.003">10.1016/j.csbj.2018.01.003</a>.</p>
</div>
<div id="ref-Zenodo">
<p>“Zenodo - Research. Shared.” <a href="https://zenodo.org/" class="uri">https://zenodo.org/</a>.</p>
</div>
<div id="ref-Zhao2012">
<p>Zhao, Jun, Jose Manuel Gomez-Perez, Khalid Belhajjame, Graham Klyne, Esteban Garcia-Cuesta, Aleix Garrido, Kristina Hettne, Marco Roos, David De Roure, and Carole Goble. 2012. “Why Workflows Break Understanding and Combating Decay in Taverna Workflows.” In <em>2012 IEEE 8th International Conference on E-Science</em>. IEEE. doi:<a href="https://doi.org/10.1109/escience.2012.6404482">10.1109/escience.2012.6404482</a>.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>For the source code that support this work we have used a similar publishing feature with Zenodo.<a href="#fnref1">↩</a></p></li>
<li id="fn2"><p><a href="https://www.commonwl.org/#Implementations" class="uri">https://www.commonwl.org/#Implementations</a><a href="#fnref2">↩</a></p></li>
<li id="fn3"><p>The CWL 1.1 specification will add a hint <code>WorkReuse</code> for this purpose.<a href="#fnref3">↩</a></p></li>
<li id="fn4"><p>See for instance <a href="https://view.commonwl.org/about#format" class="uri">https://view.commonwl.org/about#format</a><a href="#fnref4">↩</a></p></li>
</ol>
</div>
